<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>pgxBioc2022Workshop • pgxBioc2022Workshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="pgxBioc2022Workshop">
<meta property="og:description" content="This workshop will introduce users to the CoreGx and PharmacoGx R
    packages, which are useful tools for pharmacogenomic modelling to discover
    biomarkers of treatment response in cancer model systems. PharmacoGx
    specifically focuses on drug sensitivity experiments in cancer cell lines,
    which will be the major focus of this workshop. Recent updates to PharmacoGx have
    added support for analyzing drug combination experiments to discover molecular biomarkers
    of drug synergy or antagonism. In this workshop users will learn how to use CoreGx and
    PharmacoGx to build a PharmacoSet object from cancer cell-lines with both their molecular
    phenotype and response to drug combination therapy characterized. This object will then be used
    to compute common metrics of drug synergy and antagonism and to use available molecular
    data to discover genes associated with response or resistance to both mono and drug combination
    therapies.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">pgxBioc2022Workshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/PGxWorkshop.html">Pharmacogenomic Analysis of Drug Combination Experiments to Identify Biomarkers of Response or Resistance</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://bhklab.github.io/bioc2022Workshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="pharmacogenomic-analysis-of-drug-combination-experiments-to-identify-biomarkers-of-response-or-resistance">Pharmacogenomic Analysis of Drug Combination Experiments to Identify Biomarkers of Response or Resistance<a class="anchor" aria-label="anchor" href="#pharmacogenomic-analysis-of-drug-combination-experiments-to-identify-biomarkers-of-response-or-resistance"></a>
</h1></div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="instructors-names-and-contact-information">Instructor(s) name(s) and contact information<a class="anchor" aria-label="anchor" href="#instructors-names-and-contact-information"></a>
</h2>
<ul>
<li>Christopher Eeles <a href="mailto:christopher.eeles@uhnresearch.ca" class="email">christopher.eeles@uhnresearch.ca</a>
</li>
<li>Petr Smirnov <a href="mailto:petr.smirnov@utoronto.ca" class="email">petr.smirnov@utoronto.ca</a>
</li>
<li>Benjamin Haibe-Kains <a href="mailto:benjamin.haibe.kains@utoronto.ca" class="email">benjamin.haibe.kains@utoronto.ca</a>
</li>
</ul>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="workshop-description">Workshop Description<a class="anchor" aria-label="anchor" href="#workshop-description"></a>
</h2>
<!-- Alias package links -->
<p><code>{r eval=TRUE, echo=FALSE} PGx &lt;- BiocStyle::Biocpkg("PharmacoGx") CGx &lt;- BiocStyle::Biocpkg("CoreGx") TGx &lt;- BiocStyle::Biocpkg("ToxicoGx") RGx &lt;- BiocStyle::Biocpkg("RadioGx") Xv &lt;- BiocStyle::Biocpkg("Xeva") DT &lt;- BiocStyle::CRANpkg("data.table") SE &lt;- BiocStyle::Biocpkg("SummarizedExperiment") MAE &lt;- BiocStyle::Biocpkg("MultiAssayExperiment")</code></p>
<p>This workshop will introduce users to the <code>r CGx</code> and <code>r PGx</code> R packages, which are useful tools for pharmacogenomic modelling to discover biomarkers of treatment response in cancer model systems. <code>r PGx</code> specifically focuses on drug sensitivity experiments in cancer cell lines, which will be the major focus of this workshop. Additional infrastructure from our lab includes <code>r TGx</code> for toxicogenomics in healthy human cell-lines, <code>r RGx</code> for radiogenomics in cancer cell-lines and <code>r Xv</code> for pharmacogenomics in patient derived xenograph (PDX) murine models.</p>
<p>Participants will learn the fundamentals of using <code>r CGx</code> and <code>r PGx</code> to create a <code>PharmacoSet</code>—an integrative container for the storage, analysis and visualization of pharmacogenomic experiments. Particular focus will be placed on newly developed support for storing, analyzing and visualizing drug combination sensitivity experiments and correlating results therefrom with multi-omic molecular profiles to discover biomarkers of drug sensitivity, resistance, synergy, or antagonism.</p>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Knowledge of or interest in pharmacogenomics</li>
<li>Familiarity with core Bioconductor classes such as the and <code>r SE</code> and <code>r MAE</code>
</li>
<li>Curiosity about or experience with <code>r DT</code> an asset!</li>
</ul>
<p>Useful publications:</p>
<ul>
<li>Smirnov, P., Safikhani, Z., El-Hachem, N., Wang, D., She, A., Olsen, C., Freeman, M., Selby, H., Gendoo, D. M. A., Grossmann, P., Beck, A. H., Aerts, H. J. W. L., Lupien, M., Goldenberg, A., &amp; Haibe-Kains, B. (2016). PharmacoGx: An R package for analysis of large pharmacogenomic datasets. Bioinformatics (Oxford, England), 32(8), 1244–1246. <a href="https://doi.org/10.1093/bioinformatics/btv723" class="external-link uri">https://doi.org/10.1093/bioinformatics/btv723</a>
</li>
<li>Tonekaboni, M., Ali, S., Soltan Ghoraie, L., Manem, V. S. K. &amp; Haibe-Kains, B. Predictive approaches for drug combination discovery in cancer. Brief Bioinform 19, 263–276 (2018).</li>
</ul>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="workshop-participation">Workshop Participation<a class="anchor" aria-label="anchor" href="#workshop-participation"></a>
</h3>
<!-- The workshop will be presented as a set of analysis steps to be replicated by the learners, with instructors
available to explain the why and how of applying these functions to the given datasets. Learners will write analysis
scripts as well as use interactive commands to explore the generated data structures and results. Learners will then
brainstorm potential applications of the analysis results in their field as well as comment on use case examples
presented by the instructors on research in our lab. -->
<p>Participants expected to have the following required packages installed on their machines to be able to run the commands along with the instructors:</p>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="r--bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r--bioconductor-packages-used"></a>
</h3>
<p>Bioconductor:</p>
<ul>
<li><code>r CGx</code></li>
<li><code>r PGx</code></li>
</ul>
<p>CRAN:</p>
<ul>
<li><code>r DT</code></li>
<li><code>r BiocStyle::CRANpkg("ggplot2")</code></li>
</ul>
</div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h2>
<p>For a 1.5-hr workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction to CoreGx and PharmacoGx</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Overview of Data Structures</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>How the <code>TRE</code> Supports Drug Combinations Data Analysis</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Using a <code>DataMapper</code> to build a Drug Combo PharmacoSet</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Dose Response Models and Drug Sensitivity measures</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Drug Combination Synergy Models</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Biomarker Discovery</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Introduction to Resources for Biomarker Validation</td>
<td>5m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h2>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h3>
<ul>
<li>Describe pharmacogenomic mono and combination datasets and their usefulness in cancer research</li>
<li>Understand how experimental designs and research questions map onto data structures</li>
<li>Learn how to extract information from these datasets</li>
<li>Learn how to visualize experimental results from these datasets</li>
<li>Learn how to model dose-response for both monotherapy and combination small compound datasets</li>
<li>Learn measures to quantify response and synergy in cell-line sensitivity screens</li>
</ul>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>Describe use cases for <code>r CGx</code> and <code>r PGx</code> in pharmacogenomics and beyond</li>
<li>Understand the structure of the <code>CoreSet</code> and <code>PharmacoSet</code> classes to facilitate their use in subsequent analyses</li>
<li>Learn how the <code>TreatmentResponseExperiment</code> provides a highly flexible container for storing high-throughput drug combination screening data</li>
<li>Use a <code>TREDataMapper</code> to create a <code>TreatmentResponseExperiment</code> and incorporate it into a <code>PharmacoSet</code>
</li>
<li>Learn to apply arbitrary R functions over a <code>TreatmentResponseExperiment</code> using the <code>aggregate</code> method</li>
<li>Access the molecular features, dose-response and metadata contained within the data structures defined in our packages</li>
<li>Fit Hill Slope models to dose-response experiments using small compound treatments in cell-lines</li>
<li>Calculate the AAC, AUC, and IC50 metrics for response quantification in cell lines</li>
<li>Compute and store various metrics of drug synergy and antagonism using <code>r PGx</code> and the <code>PharmacoSet</code>
</li>
<li>Learn how to use <code>PharmacoGx</code> to discover univariate biomarkers of drug response, resistance, synergy, or antagnoism</li>
</ul>
<hr>
</div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="outline">Outline<a class="anchor" aria-label="anchor" href="#outline"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Introduction to CoreGx and PharmacoGx</li>
<li>Overview of Data Structures
<ol style="list-style-type: decimal">
<li>CoreSet and PharmacoSet</li>
<li>MutliAssayExperiment and SummarizedExperiment</li>
<li>TreatmentResponseExperiment</li>
</ol>
</li>
<li>How TRE Support Drug Combinations Data Analysis</li>
<li>Using Data Mapper to build a Drug Combo PSet
<ol style="list-style-type: decimal">
<li>Data Mapper</li>
<li>Combining with Omics Data into a PharmacoSet</li>
</ol>
</li>
<li>Dose Response Models and Drug Sensitivity Measures</li>
<li>Drug Combination Synergy Models</li>
<li>Biomarker Discovery:
<ol style="list-style-type: decimal">
<li>Monotherapy response</li>
<li>Combination Synergy</li>
</ol>
</li>
<li>Introduction to Resources for Biomarker Validation</li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing pgxBioc2022Workshop</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Christopher Eeles <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-9962-7385" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Petr Smirnov <br><small class="roles"> Author </small>  </li>
<li>Feifei Li <br><small class="roles"> Author </small>  </li>
<li>Benjamin Haibe-Kains <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Christopher Eeles, Petr Smirnov, Feifei Li, Benjamin Haibe-Kains.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
