<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pharmacogenomic Analysis of Drug Combination Experiments to Identify Biomarkers of Response or Resistance • pgxBioc2022Workshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pharmacogenomic Analysis of Drug Combination Experiments to Identify Biomarkers of Response or Resistance">
<meta property="og:description" content="pgxBioc2022Workshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pgxBioc2022Workshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PGxWorkshop.html">Pharmacogenomic Analysis of Drug Combination Experiments to Identify Biomarkers of Response or Resistance</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://bhklab.github.io/bioc2022Workshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pharmacogenomic Analysis of Drug Combination
Experiments to Identify Biomarkers of Response or Resistance</h1>
                        <h4 data-toc-skip class="author">Christopher
Eeles</h4>
            <address class="author_afil">
      Princess Margaret Cancer
Centre<br><a class="author_email" href="mailto:#"></a><a href="mailto:christopher.eeles@uhnresearch.ca" class="email">christopher.eeles@uhnresearch.ca</a>
      </address>
                              <h4 data-toc-skip class="author">Petr
Smirnov</h4>
            <address class="author_afil">
      Princess Margaret Cancer CentreUniversity of
Toronto<br><a class="author_email" href="mailto:#"></a><a href="mailto:petr.smirnov@utoronto.ca" class="email">petr.smirnov@utoronto.ca</a>
      </address>
                              <h4 data-toc-skip class="author">Feifei
Li</h4>
            <address class="author_afil">
      University of
Toronto<br><a class="author_email" href="mailto:#"></a><a href="mailto:ff.li@mail.utoronto.ca" class="email">ff.li@mail.utoronto.ca</a>
      </address>
                  
      
      
      <div class="hidden name"><code>PGxWorkshop.Rmd</code></div>

    </div>

    
    
<!-- Remove {-} from header to show numbering -->
<div class="section level2">
<h2 class="unnumbered" id="workshop-overview">Workshop Overview<a class="anchor" aria-label="anchor" href="#workshop-overview"></a>
</h2>
<div class="section level3">
<h3 class="unnumbered" id="instructors-names-and-contact-information">Instructor(s) name(s) and contact
information<a class="anchor" aria-label="anchor" href="#instructors-names-and-contact-information"></a>
</h3>
<ul>
<li>Christopher Eeles <a href="mailto:christopher.eeles@uhnresearch.ca" class="email">christopher.eeles@uhnresearch.ca</a>
</li>
<li>Petr Smirnov <a href="mailto:petr.smirnov@utoronto.ca" class="email">petr.smirnov@utoronto.ca</a>
</li>
<li>Benjamin Haibe-Kains <a href="mailto:benjamin.haibe.kains@utoronto.ca" class="email">benjamin.haibe.kains@utoronto.ca</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 class="unnumbered" id="workshop-description">Workshop Description<a class="anchor" aria-label="anchor" href="#workshop-description"></a>
</h3>
<!-- Alias package links -->
<p>This workshop will introduce users to the <em><a href="https://bioconductor.org/packages/3.16/CoreGx" class="external-link">CoreGx</a></em> and
<em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
R packages, which are useful tools for pharmacogenomic modelling to
discover biomarkers of treatment response in cancer model systems.
<em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
specifically focuses on drug sensitivity experiments in cancer cell
lines, which will be the major focus of this workshop. Additional
infrastructure from our lab includes <em><a href="https://bioconductor.org/packages/3.16/ToxicoGx" class="external-link">ToxicoGx</a></em>
for toxicogenomics in healthy human cell-lines, <em><a href="https://bioconductor.org/packages/3.16/RadioGx" class="external-link">RadioGx</a></em>
for radiogenomics in cancer cell-lines and <em><a href="https://bioconductor.org/packages/3.16/Xeva" class="external-link">Xeva</a></em> for
pharmacogenomics in patient derived xenograph (PDX) murine models.</p>
<p>Participants will learn the fundamentals of using <em><a href="https://bioconductor.org/packages/3.16/CoreGx" class="external-link">CoreGx</a></em> and
<em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
to create a <code>PharmacoSet</code>—an integrative container for the
storage, analysis and visualization of pharmacogenomic experiments.
Particular focus will be placed on newly developed support for storing,
analyzing and visualizing drug combination sensitivity experiments and
correlating results therefrom with multi-omic molecular profiles to
discover biomarkers of drug sensitivity, resistance, synergy, or
antagonism.</p>
<div class="section level4">
<h4 class="unnumbered" id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h4>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Knowledge of or interest in pharmacogenomics</li>
<li>Familiarity with core Bioconductor classes such as the and <em><a href="https://bioconductor.org/packages/3.16/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
and <em><a href="https://bioconductor.org/packages/3.16/MultiAssayExperiment" class="external-link">MultiAssayExperiment</a></em>
</li>
<li>Curiosity about or experience with <em><a href="https://CRAN.R-project.org/package=data.table" class="external-link">data.table</a></em>
an asset!</li>
</ul>
<p>Useful publications:</p>
<ul>
<li>Smirnov, P., Safikhani, Z., El-Hachem, N., Wang, D., She, A., Olsen,
C., Freeman, M., Selby, H., Gendoo, D. M. A., Grossmann, P., Beck, A.
H., Aerts, H. J. W. L., Lupien, M., Goldenberg, A., &amp; Haibe-Kains,
B. (2016). PharmacoGx: An R package for analysis of large
pharmacogenomic datasets. Bioinformatics (Oxford, England), 32(8),
1244–1246. <a href="https://doi.org/10.1093/bioinformatics/btv723" class="external-link uri">https://doi.org/10.1093/bioinformatics/btv723</a>
</li>
<li>Tonekaboni, M., Ali, S., Soltan Ghoraie, L., Manem, V. S. K. &amp;
Haibe-Kains, B. Predictive approaches for drug combination discovery in
cancer. Brief Bioinform 19, 263–276 (2018).</li>
</ul>
</div>
<div class="section level4">
<h4 class="unnumbered" id="workshop-participation">Workshop Participation<a class="anchor" aria-label="anchor" href="#workshop-participation"></a>
</h4>
<!-- The workshop will be presented as a set of analysis steps to be replicated by the learners, with instructors
available to explain the why and how of applying these functions to the given datasets. Learners will write analysis
scripts as well as use interactive commands to explore the generated data structures and results. Learners will then
brainstorm potential applications of the analysis results in their field as well as comment on use case examples
presented by the instructors on research in our lab. -->
<p>Participants expected to have the following required packages
installed on their machines to be able to run the commands along with
the instructors:</p>
</div>
<div class="section level4">
<h4 class="unnumbered" id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages
used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h4>
<p>Bioconductor:</p>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.16/CoreGx" class="external-link">CoreGx</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em></li>
</ul>
<p>CRAN:</p>
<ul>
<li><em><a href="https://CRAN.R-project.org/package=data.table" class="external-link">data.table</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">ggplot2</a></em></li>
</ul>
</div>
</div>
<div class="section level3">
<h3 class="unnumbered" id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<p>For a 1.5-hr workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction to CoreGx and PharmacoGx</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Overview of Data Structures</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>How the <code>TRE</code> Supports Drug Combinations Data
Analysis</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Using a <code>DataMapper</code> to build a Drug Combo
PharmacoSet</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Dose Response Models and Drug Sensitivity measures</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Drug Combination Synergy Models</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Biomarker Discovery</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Further Resources</td>
<td>5m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 class="unnumbered" id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
<div class="section level4">
<h4 class="unnumbered" id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h4>
<ul>
<li>Describe pharmacogenomic mono and combination datasets and their
usefulness in cancer research</li>
<li>Understand how experimental designs and research questions map onto
data structures</li>
<li>Learn how to extract information from these datasets</li>
<li>Learn how to visualize experimental results from these datasets</li>
<li>Learn how to model dose-response for both monotherapy and
combination small compound datasets</li>
<li>Learn measures to quantify response and synergy in cancer cell-line
sensitivity screens</li>
</ul>
</div>
<div class="section level4">
<h4 class="unnumbered" id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li>Describe use cases for <em><a href="https://bioconductor.org/packages/3.16/CoreGx" class="external-link">CoreGx</a></em> and
<em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
in pharmacogenomics and beyond</li>
<li>Understand the structure of the <code>CoreSet</code> and
<code>PharmacoSet</code> classes to facilitate their use in subsequent
analyses</li>
<li>Learn how the <code>TreatmentResponseExperiment</code> provides a
highly flexible container for storing high-throughput drug combination
screening data</li>
<li>Use a <code>TREDataMapper</code> to create a
<code>TreatmentResponseExperiment</code> and incorporate it into a
<code>PharmacoSet</code>
</li>
<li>Learn to apply arbitrary R functions over a
<code>TreatmentResponseExperiment</code> using the
<code>aggregate</code> and <code>endoaggregate</code> methods</li>
<li>Access the molecular features, dose-response and metadata contained
within the data structures defined in our packages</li>
<li>Fit Hill Slope models to dose-response experiments using small
compound treatments in cell-lines</li>
<li>Calculate the AAC, AUC, and IC50 metrics for response quantification
in cell lines</li>
<li>Compute and store various metrics of drug synergy and antagonism
using <em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
and the <code>TreatmentResponseExperiment</code>
</li>
<li>Learn how to use <code>PharmacoGx</code> to discover univariate
biomarkers of drug response, resistance, synergy, or antagnoism</li>
</ul>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="introduction-to-coregx-and-pharmacogx">Introduction to CoreGx and PharmacoGx<a class="anchor" aria-label="anchor" href="#introduction-to-coregx-and-pharmacogx"></a>
</h2>
<p>This tutorial, titled <em>Pharmacogenomic Analysis of Drug
Combination Experiments to Identify Biomarkers of Response or
Resistance</em>, focuses on using the <em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
R package to correlate treatment response, measured as the viability of
cancer cell-lines after <em>in vitro</em> drug treatment, with their
respective multi-omic profiles. <em><a href="https://bioconductor.org/packages/3.16/CoreGx" class="external-link">CoreGx</a></em>
provides the core infrastructure for storing, analyzing, and visualizing
generic treatment response experiments. It provides methods and classes
which can be inherited from in downstream packages, such as <em><a href="https://bioconductor.org/packages/3.16/ToxicoGx" class="external-link">ToxicoGx</a></em>
and <em><a href="https://bioconductor.org/packages/3.16/RadioGx" class="external-link">RadioGx</a></em>.
We hope that the <code>CoreSet</code> object is generalized enough that
it can be reused by other developers for their specific treatment (or
stimuli) response use case.</p>
<div class="section level3">
<h3 id="coregx">CoreGx<a class="anchor" aria-label="anchor" href="#coregx"></a>
</h3>
<div class="section level4">
<h4 id="package-nomenclature">Package Nomenclature<a class="anchor" aria-label="anchor" href="#package-nomenclature"></a>
</h4>
<p>To facilitate modularization of the GxSuite of R packages, we have
shifted the nomenclature within a <code>CoreSet</code>—and therefore in
inheriting packages —to be more general.</p>
<p>To this end, have made the following changes:</p>
<ul>
<li>Previous reference to cell (cell-line) have become sample, allowing
the <code>CoreSet</code> to be used for other model systems</li>
<li>Drug (radiation in <code>RadioGx</code>) have become treatment,
allowing the <code>CoreSet</code> to be treatment type (or stimuli)
agnostic</li>
<li>Sensitivity will become response (sensitivity slot becomes
treatmentResponse)</li>
</ul>
<p>As a result of these changes, the names of some common accessors have
been updated. The old accessors still remain functional to ensure
backwards compatibility for at least two Bioconductor releases. A
deprecation warning will be added to old accessors informing users of
the corresponding new function, as per Bioconductor best practices.</p>
</div>
</div>
<div class="section level3">
<h3 id="pharmacogx">PharmacoGx<a class="anchor" aria-label="anchor" href="#pharmacogx"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
stores drug screening data together with molecular profiling of
cell-lines in an object called a <code>PharmacoSet</code>, or
<code>PSet</code> for short. This object inherits for the
<code>CoreSet</code> class defined in our package <code>CoreGx</code>,
which is used to share common datas structure and method across our
suite of package. The primary use case for <code>PharmacoGx</code> is to
provide a standardized and highly curated container for high-throughput
screens in cancer-cell lines, with the goal of enabling discovery of
biomarkers of treatment response or resistance which can be easily
compared and validated across large published pharmacogneomic
datasets.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="overview-of-data-structures">Overview of Data Structures<a class="anchor" aria-label="anchor" href="#overview-of-data-structures"></a>
</h2>
<p>The GxSuite of packages make use of various Bioconductor classes for
storing molecular profile data.</p>
<div class="section level3">
<h3 id="coreset-and-pharmacoset">
<code>CoreSet</code> and <code>PharmacoSet</code><a class="anchor" aria-label="anchor" href="#coreset-and-pharmacoset"></a>
</h3>
<p>With the recent updates in the structure and nomenclature of the
<code>CoreSet</code>, the underling data structures are essentially
identical. Previously, the “treatment” slot was specific to inheriting
classes and named according to their domain specific use case, for
example drug in <em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
and radiation in <em><a href="https://bioconductor.org/packages/3.16/RadioGx" class="external-link">RadioGx</a></em>.
Correspondingly, the “sample” slot was named cell and corresponded to
<em>in vitro</em> immortalized human cancer cell-lines.</p>
<div class="figure">
<img src="CoreSet-class.drawio.png" alt=""><p class="caption"><strong>CoreSet / PharmacoSet class diagram</strong>.
Objects comprising a <code>CoreSet</code> are enclosed in boxes. The
first box indicates the type and name of each object. The second box
indicates the structure of an object or class. The third box shows
accessor methods from <code>PharmacoGx</code> for that specific object.
‘=&gt;’ represents return and specifies what is returned from that item
or method.</p>
</div>
</div>
<div class="section level3">
<h3 id="multiassayexperiment-and-summarizedexperiment">
<code>MultiAssayExperiment</code> and
<code>SummarizedExperiment</code><a class="anchor" aria-label="anchor" href="#multiassayexperiment-and-summarizedexperiment"></a>
</h3>
<p>The <code>MultiAssayExperiment</code> (<code>MAE</code>) is a common
Bioconductor class for storing the results of diverse multi-omic
profiling experiments, including mutation, RNA expression, copy number
abberation and protein abundace. Along with the
<code>SummarizedExperiment</code> (<code>SE</code>), this object is used
to store molecular phenotype data within a <code>CoreSet</code> and
<code>PharmacoSet</code>.</p>
<p>At the heart of both objects is a feature by sample matrix, with the
major difference being that the <code>SE</code> require congruent
matrices (i.e., observations on the same set of features and samples)
where as the <code>MAE</code> allows storage of disjoint sets of feature
and samples via the use of the <code>sampleMap</code>, which,
conceptually, stores a graph of the mapping from samples to each assay
in the object.</p>
<div class="figure">
<img src="MultiAssayExperiment_SummarizedExperiment_schematics.png" alt=""><p class="caption"><strong>SummarizedExperiment</strong> and
<strong>MultiAssayExperiment</strong> classes.</p>
</div>
<p>For both these objects, rows map to molecular features (often genomic
coordinates to better generalize across diverse molecular datatypes),
and columns to samples. Metadata for both is available via the
<code>rowData</code> and <code>colData</code> accessors,
respectively.</p>
</div>
<div class="section level3">
<h3 id="treatmentresponseexperiment">
<code>TreatmentResponseExperiment</code><a class="anchor" aria-label="anchor" href="#treatmentresponseexperiment"></a>
</h3>
<p>The <code>TRE</code> object shares much of its design with the
<code>MultiAssayExperiment</code> and <code>SummarizedExperiment</code>.
We store dimension metadata in <code>rowData</code> and
<code>colData</code>, similar to an <code>SE</code> but also allow
disjoint rows and columns to be stored in <code>assays</code>, like an
<code>MAE</code>. The major difference is that instead of using a matrix
or array for storing assay observations, we instead store them as a long
format <code>data.table</code> where integer keys define the relevant
“shape” of the underlying data and observations live within columns of
that table.</p>
<p>For those familar with relational modelling from RDBMS systems, our
design has taken inspiration from the star schema often used in OLAP
applications to improve the efficiency of analytic database queries.</p>
<div class="figure">
<img src="TreatmentResponseExperimentClassDiagram.drawio.png" alt=""><p class="caption"><strong>TreatmentResponseExperiment class
diagram</strong>. Objects comprising a
<code>TreatmentResponseExperiment</code> are enclosed in boxes. The
first box indicates the type and name of each object. The second box
indicates the structure of an object or class. The third box shows
accessor methods from <code>PharmacoGx</code> for that specific object.
‘=&gt;’ represents return and specifies what is returned from that item
or method.</p>
</div>
<p>The <code>TRE</code> differentiates between full assays and summary
assays by the presence of various identifier columns which can be viewed
with the <code>idCols</code> method. Each <code>TRE</code> has a set of
<code>rowIDs</code> and <code>colIDs</code> which define these
respective dimensions. By storing an <code>assayIndex</code>, analagous
to the <code>sampleMap</code> from the <code>MAE</code> object, we are
able to summarize over one or more row or column identifiers while
storing only the relevant rows within the assays slot, allowing for a
measure of compression relative to a matrix or array, where missing
values must be stored as <code>NA</code>.</p>
<p>The association between the various identifiers in the object are
shown below, with the <code>assayIndex</code> being the central
repository for the graph like relationships between the
<code>rowIDs</code>, <code>colIDs</code> and each respective
<code>assay</code>.</p>
<div class="figure">
<img src="TreatmentResponseExperimentStructure.drawio.png" alt=""><p class="caption">The structure of a
<code>TreatmentResponseExperiment</code> centers around the
<code>assayIndex</code> which contains a mapping from unique identifiers
for row or columns and the primary key of each assay in the object.
Summaries over rowKey and colKeys, and thus their corresponding
identifiers, and made possible by repeating a givan assay key for ecah
rowKey/colKey combination it has been summarized over.</p>
</div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="how-the-tre-support-drug-combinations-data-analysis">How the <code>TRE</code> Support Drug Combinations Data
Analysis<a class="anchor" aria-label="anchor" href="#how-the-tre-support-drug-combinations-data-analysis"></a>
</h2>
<div class="section level3">
<h3 id="drug-combination-experiments">Drug Combination Experiments<a class="anchor" aria-label="anchor" href="#drug-combination-experiments"></a>
</h3>
<p>The field of precision oncology is increasingly interested in
exploring possible synergies between cancer therapies as a means of
overcoming acquired treatment resistance or overcoming in complete
responses to first line therapies. To accelerate the process of
translating insights from <em>in vitro</em> drug combination experiments
into clinical practice, we have extended the <code>CoreSet</code> and
<code>PharmacoSet</code> classes with a new object, the
<code>TreatmentResponseExperiment</code> (<code>TRE</code>), which has
been specifically designed to handle high dimensional biological
stimilus-response data.</p>
<p>While drug combinations are the first use case for this data
structure, as cancer model systems continue to advance we expected to
also see sample-wise combinations. For example, 3D organoid models could
include multiple cell-lines or patient tissue types. Additionally, we
can imagine the object being useful for other types of stimuli-response
experiments across the biological sciences, such as monitoring the
response of nerve cells to various types of sensory stimuli.</p>
</div>
<div class="section level3">
<h3 id="building-a-drug-combination-pharmacoset">Building A Drug Combination PharmacoSet<a class="anchor" aria-label="anchor" href="#building-a-drug-combination-pharmacoset"></a>
</h3>
<div class="section level4">
<h4 id="datamapper-class">
<code>DataMapper</code> Class<a class="anchor" aria-label="anchor" href="#datamapper-class"></a>
</h4>
<p>The <code>DataMapper</code> is an abstract class included in
<code>CoreGx</code> which defines metadata specifying a “recipe” to go
from rawdata files associated with a specific experiment or publication
to a curated and standardized data format within an R <code>S4</code>
class.</p>
<p>In <em><a href="https://bioconductor.org/packages/3.16/CoreGx" class="external-link">CoreGx</a></em> we
have implemented the <code>TREDataMapper</code> class, which encodes how
the rawdata from a drug combination experiment should be transformed
into its associated <code>TRE</code> object. A primary motivation for
the creation of this class was to document and standardize creation of
<code>TRE</code> objects from the myriad experimental designs available
in drug combination experiments in the wild. It is our hope that, for
example, 3x3 drug combination experiments have enough in common that we
could theoretically create a <code>3by3TREDataMapper</code> which works
for (almost) all publications following this design.</p>
<p>As things stand right now, the <code>TREDataMapper</code> primarily
acts as helper class to explore the specifics of a drug combination
experimental design. It allows us to generate hypotheses about which
identifiers are required to identify observations in an assay. Below we
will see how, along with the <code>guessMapping</code> method, we can
use the <code>TREDataMapper</code> to ensure we have mapped all data
from the NCI-ALMANAC drug combination experiment into a <code>TRE</code>
object, ensuring that all study observations and relevant metadata are
captured in our <code>TRE</code> and therefore in our
<code>PharmacoSet</code>.</p>
<p>By curating and releasing such <code>PharmacoSet</code> objects, we
ensure that our users can access all information from the original
publication via <em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
and therefore easily ensure the reproducibility of the results as well
as simplify their comparison against other curated mono and combination
therapy <code>PharmacoSet</code>s object. In doing so, we help to
improve the reproducibility and consistency of <em>in vitro</em>
pharmacogenomic experiments and accelerate translation of insights from
the lab into the clinic.</p>
</div>
<div class="section level4">
<h4 id="using-the-tredatamapper-to-construct-a-treatmentresponseexperiment">Using the <code>TREDataMapper</code> to Construct a
<code>TreatmentResponseExperiment</code><a class="anchor" aria-label="anchor" href="#using-the-tredatamapper-to-construct-a-treatmentresponseexperiment"></a>
</h4>
<p>For this workshop we have included a subset of the NCI_ALMANAC_2017
<code>PharmacoSet</code>, curated from <span class="citation">(Holbeck
et al. 2017)</span>. When approaching new treament response experiment
data, the first step is to merge it into a single, long format table by
attaching additional metadata for treatments, sample and any other
entities defined in the publication. We have done this already and
included it as the <code>NCI_ALMANAC_raw</code> dataset in the workshop
package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">NCI_ALMANAC_raw</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">NCI_ALMANAC_raw</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="10%">
<col width="3%">
<col width="4%">
<col width="4%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="7%">
<col width="2%">
<col width="2%">
<col width="3%">
<col width="4%">
<col width="4%">
<col width="2%">
<col width="3%">
<col width="2%">
<col width="3%">
<col width="3%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="3%">
<col width="2%">
<col width="3%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="2%">
</colgroup>
<thead><tr class="header">
<th align="left">treatment1id</th>
<th align="left">treatment2id</th>
<th align="right">treatment1dose</th>
<th align="right">treatment2dose</th>
<th align="left">sampleid</th>
<th align="right">CONC1</th>
<th align="right">CONC2</th>
<th align="right">NSC2</th>
<th align="left">CELLNAME</th>
<th align="right">CELLNBR</th>
<th align="left">PANEL</th>
<th align="right">PANELNBR</th>
<th align="right">viability</th>
<th align="right">PERCENTGROWTH</th>
<th align="right">PERCENTGROWTHNOTZ</th>
<th align="right">EXPECTEDGROWTH</th>
<th align="right">TESTVALUE</th>
<th align="right">CONTROLVALUE</th>
<th align="right">TZVALUE</th>
<th align="right">CONCINDEX1</th>
<th align="right">CONCINDEX2</th>
<th align="right">SAMPLE1</th>
<th align="right">SAMPLE2</th>
<th align="right">SCORE</th>
<th align="right">COMBODRUGSEQ</th>
<th align="left">STUDY</th>
<th align="left">TESTDATE</th>
<th align="left">PLATE</th>
<th align="left">SCREENER</th>
<th align="left">PREFIX2</th>
<th align="right">bio_rep</th>
<th align="right">tech_rep</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1,3-Bis(2-chloroethyl)-1-nitrosourea</td>
<td align="left"></td>
<td align="right">0.05</td>
<td align="right">NA</td>
<td align="left">786-O</td>
<td align="right">0</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">786-0</td>
<td align="right">18</td>
<td align="left">Renal Cancer</td>
<td align="right">9</td>
<td align="right">98.89562</td>
<td align="right">98.367</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">2.776</td>
<td align="right">2.807</td>
<td align="right">0.909</td>
<td align="right">-1</td>
<td align="right">0</td>
<td align="right">23</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">5587379</td>
<td align="left">1307AC31</td>
<td align="left">07/08/2013</td>
<td align="left">3577</td>
<td align="left">1A</td>
<td align="left"></td>
<td align="right">1</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">1,3-Bis(2-chloroethyl)-1-nitrosourea</td>
<td align="left"></td>
<td align="right">0.05</td>
<td align="right">NA</td>
<td align="left">A-498</td>
<td align="right">0</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">A498</td>
<td align="right">13</td>
<td align="left">Renal Cancer</td>
<td align="right">9</td>
<td align="right">98.63179</td>
<td align="right">96.232</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">2.451</td>
<td align="right">2.485</td>
<td align="right">1.589</td>
<td align="right">-1</td>
<td align="right">0</td>
<td align="right">23</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">5587364</td>
<td align="left">1307AC31</td>
<td align="left">07/08/2013</td>
<td align="left">3432</td>
<td align="left">1A</td>
<td align="left"></td>
<td align="right">1</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">1,3-Bis(2-chloroethyl)-1-nitrosourea</td>
<td align="left"></td>
<td align="right">0.05</td>
<td align="right">NA</td>
<td align="left">A-549</td>
<td align="right">0</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">A549/ATCC</td>
<td align="right">4</td>
<td align="left">Non-Small Cell Lung Cancer</td>
<td align="right">1</td>
<td align="right">100.91533</td>
<td align="right">101.089</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">1.764</td>
<td align="right">1.748</td>
<td align="right">0.325</td>
<td align="right">-1</td>
<td align="right">0</td>
<td align="right">23</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">5587194</td>
<td align="left">1307AC31</td>
<td align="left">07/08/2013</td>
<td align="left">3548</td>
<td align="left">1A</td>
<td align="left"></td>
<td align="right">1</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">1,3-Bis(2-chloroethyl)-1-nitrosourea</td>
<td align="left"></td>
<td align="right">0.05</td>
<td align="right">NA</td>
<td align="left">ACHN</td>
<td align="right">0</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">ACHN</td>
<td align="right">23</td>
<td align="left">Renal Cancer</td>
<td align="right">9</td>
<td align="right">100.81652</td>
<td align="right">101.054</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">2.099</td>
<td align="right">2.082</td>
<td align="right">0.492</td>
<td align="right">-1</td>
<td align="right">0</td>
<td align="right">23</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">5587384</td>
<td align="left">1307AC31</td>
<td align="left">07/08/2013</td>
<td align="left">3606</td>
<td align="left">1A</td>
<td align="left"></td>
<td align="right">1</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">1,3-Bis(2-chloroethyl)-1-nitrosourea</td>
<td align="left"></td>
<td align="right">0.05</td>
<td align="right">NA</td>
<td align="left">BT-549</td>
<td align="right">0</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="left">BT-549</td>
<td align="right">13</td>
<td align="left">Breast Cancer</td>
<td align="right">5</td>
<td align="right">95.67643</td>
<td align="right">91.405</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">2.058</td>
<td align="right">2.151</td>
<td align="right">1.069</td>
<td align="right">-1</td>
<td align="right">0</td>
<td align="right">23</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">5587284</td>
<td align="left">1307AC31</td>
<td align="left">07/08/2013</td>
<td align="left">3577</td>
<td align="left">1A</td>
<td align="left"></td>
<td align="right">1</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
<p>Viewing the table, we see that each column represents a variable, and
that the number of rows in the table corresponds to the number of
dose-response observations measured in the <code>NCI-ALMANAC</code>
study. To those familiar with the R <code>tidyverse</code>, you may
recognize that this dataset is in <a href="https://r4ds.had.co.nz/tidy-data.html" class="external-link">tidy format</a>. It is also
said be in a “long” format because all observations are contained in a
single column; contrast this with a “wide” format where columns
generally respresent samples and values are observations are stored
across both rows and columns (<em>i.e.</em>, as in a matrix or
tensor).</p>
<p>In order for the helper methods of the <code>TREDataMapper</code>
class to work, our rawdata must first be in a single large, tidy
<code>data.frame</code> or <code>data.table</code>. Since these
conditions are already met for the included example data, we can create
of <code>TREDataMapper</code> object and start exploring the to
dimensionality of the experiment in question!</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">treDataMapper</span> <span class="op">&lt;-</span> <span class="fu">TREDataMapper</span><span class="op">(</span>rawdata<span class="op">=</span><span class="va">NCI_ALMANAC_raw</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>##  [33m [1m [3m&lt;TREDataMapper&gt; 
##  [23m [22m [39m [33m [1mrawdata: [22m [39m dim(534179, 32)
##                              treatment1id treatment2id treatment1dose
##                                    &lt;char&gt;       &lt;char&gt;          &lt;num&gt;
##   1: 1,3-Bis(2-chloroethyl)-1-nitrosourea                        0.05
##   2: 1,3-Bis(2-chloroethyl)-1-nitrosourea                        0.05
##   3: 1,3-Bis(2-chloroethyl)-1-nitrosourea                        0.05
##   29 variables not shown: [treatment2dose &lt;num&gt;, sampleid &lt;char&gt;, CONC1
##     &lt;num&gt;, CONC2 &lt;num&gt;, NSC2 &lt;int&gt;, CELLNAME &lt;char&gt;, CELLNBR &lt;int&gt;,
##     PANEL &lt;char&gt;, PANELNBR &lt;int&gt;, viability &lt;num&gt;, ...] 
## 
 [33m [1mrowDataMap: [22m [39m
##    [32mrowIDs: [39m 
##    [32mrowMeta: [39m 
##  [33m [1mcolDataMap: [22m [39m
##    [32mcolIDs: [39m 
##    [32mcolMeta: [39m  [33m [1m
## assayMap: [22m [39m [33m [1m
## metadataMap: [22m [39m [32m NA [39m</code></pre>
<p>A <code>DataMapper</code> as a concept is simply a collection of
metadata around our raw data which defines how different columns map to
different aspects of our target <code>S4</code> class. For a
<code>TRE</code>, similar to a <em><a href="https://bioconductor.org/packages/3.16/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>,
we have the concept of <code>rowData</code> which defines what is
considered a treatment (a specific drug combination, its doses and
potential technical replicates) and <code>colData</code> which defines
what is considered a sample (a specific cell-line and possible
biological replicate identifier).</p>
<p>The <code>guessMapping</code> method can be used to test our
hypotheses about which information maps to treatments (rows), samples
(columns) or assays (experimental observations). This is done by
defining a set of groupings and querying the table to find other columns
which map 1:1 with those groups.</p>
<p>For our first guess, lets assume there are no replicates and see if
we are able to map all of our rawdata columns using just treatment,
treatment dose and sample identifiers. Once we format our guess for
rows, columns and assays, we specify the subset parmeter to indicate we
want to remove any mapped columns before moving on to the next group.
This prevents multimapping a column to several groups and is generally
recommended.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">groups1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    rowDataMap<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"treatment1dose"</span>,</span>
<span>        <span class="st">"treatment2dose"</span><span class="op">)</span>,</span>
<span>    colDataMap<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sampleid"</span><span class="op">)</span>,</span>
<span>    assayMap<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"treatment1dose"</span>,</span>
<span>        <span class="st">"treatment2dose"</span>, <span class="st">"sampleid"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">guess1</span> <span class="op">&lt;-</span> <span class="fu">guessMapping</span><span class="op">(</span><span class="va">treDataMapper</span>, groups<span class="op">=</span><span class="va">groups1</span>, subset<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [CoreGx::guessMapping,LongTableDataMapper-method]</span></span>
<span><span class="co">##  Mapping for group rowDataMap: treatment1id, treatment2id, treatment1dose, treatment2dose</span></span></code></pre>
<pre><code><span><span class="co">## [CoreGx::guessMapping,LongTableDataMapper-method]</span></span>
<span><span class="co">##  Mapping for group colDataMap: sampleid</span></span></code></pre>
<pre><code><span><span class="co">## [CoreGx::guessMapping,LongTableDataMapper-method]</span></span>
<span><span class="co">##  Mapping for group assayMap: treatment1id, treatment2id, treatment1dose, treatment2dose, sampleid</span></span></code></pre>
<pre><code><span><span class="co">## $metadata</span></span>
<span><span class="co">## $metadata$id_columns</span></span>
<span><span class="co">## [1] NA</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $metadata$mapped_columns</span></span>
<span><span class="co">## character(0)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $rowDataMap</span></span>
<span><span class="co">## $rowDataMap$id_columns</span></span>
<span><span class="co">## [1] "treatment1id"   "treatment2id"   "treatment1dose" "treatment2dose"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $rowDataMap$mapped_columns</span></span>
<span><span class="co">## [1] "CONC1"      "CONC2"      "NSC2"       "CONCINDEX2"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $colDataMap</span></span>
<span><span class="co">## $colDataMap$id_columns</span></span>
<span><span class="co">## [1] "sampleid"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $colDataMap$mapped_columns</span></span>
<span><span class="co">## [1] "CELLNAME" "CELLNBR"  "PANEL"    "PANELNBR"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $assayMap</span></span>
<span><span class="co">## $assayMap$id_columns</span></span>
<span><span class="co">## [1] "treatment1id"   "treatment2id"   "treatment1dose" "treatment2dose"</span></span>
<span><span class="co">## [5] "sampleid"      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $assayMap$mapped_columns</span></span>
<span><span class="co">## [1] "EXPECTEDGROWTH" "SAMPLE2"        "SCORE"         </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $unmapped</span></span>
<span><span class="co">##  [1] "viability"         "PERCENTGROWTH"     "PERCENTGROWTHNOTZ"</span></span>
<span><span class="co">##  [4] "TESTVALUE"         "CONTROLVALUE"      "TZVALUE"          </span></span>
<span><span class="co">##  [7] "CONCINDEX1"        "SAMPLE1"           "COMBODRUGSEQ"     </span></span>
<span><span class="co">## [10] "STUDY"             "TESTDATE"          "PLATE"            </span></span>
<span><span class="co">## [13] "SCREENER"          "PREFIX2"           "bio_rep"          </span></span>
<span><span class="co">## [16] "tech_rep"</span></span></code></pre>
<p>Observing the results of our first guess, we can see that many column
remain unmapped. This likely indicates there is an additional piece of
information needed to uniquely identify the dimensions of our drug
combination experiment. One potential cause for this is inclusion of
technical or biological replicates.</p>
<p>Indeed, for the NCI-ALMANAC both of these kinds of replicates are
present.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu">rawdata</span><span class="op">(</span><span class="va">treDataMapper</span><span class="op">)</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Classes 'data.table' and 'data.frame':   534179 obs. of  32 variables:</span></span>
<span><span class="co">##  $ treatment1id     : chr  "1,3-Bis(2-chloroethyl)-1-nitrosourea" "1,3-Bis(2-chloroethyl)-1-nitrosourea" "1,3-Bis(2-chloroethyl)-1-nitrosourea" "1,3-Bis(2-chloroethyl)-1-nitrosourea" ...</span></span>
<span><span class="co">##  $ treatment2id     : chr  "" "" "" "" ...</span></span>
<span><span class="co">##  $ treatment1dose   : num  0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 ...</span></span>
<span><span class="co">##  $ treatment2dose   : num  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">##  $ sampleid         : chr  "786-O" "A-498" "A-549" "ACHN" ...</span></span>
<span><span class="co">##  $ CONC1            : num  5e-08 5e-08 5e-08 5e-08 5e-08 5e-08 5e-08 5e-08 5e-08 5e-08 ...</span></span>
<span><span class="co">##  $ CONC2            : num  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">##  $ NSC2             : int  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">##  $ CELLNAME         : chr  "786-0" "A498" "A549/ATCC" "ACHN" ...</span></span>
<span><span class="co">##  $ CELLNBR          : int  18 13 4 23 13 3 10 15 3 2 ...</span></span>
<span><span class="co">##  $ PANEL            : chr  "Renal Cancer" "Renal Cancer" "Non-Small Cell Lung Cancer" "Renal Cancer" ...</span></span>
<span><span class="co">##  $ PANELNBR         : int  9 9 1 9 5 7 4 9 11 4 ...</span></span>
<span><span class="co">##  $ viability        : num  98.9 98.6 100.9 100.8 95.7 ...</span></span>
<span><span class="co">##  $ PERCENTGROWTH    : num  98.4 96.2 101.1 101.1 91.4 ...</span></span>
<span><span class="co">##  $ PERCENTGROWTHNOTZ: num  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">##  $ EXPECTEDGROWTH   : num  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">##  $ TESTVALUE        : num  2.78 2.45 1.76 2.1 2.06 ...</span></span>
<span><span class="co">##  $ CONTROLVALUE     : num  2.81 2.48 1.75 2.08 2.15 ...</span></span>
<span><span class="co">##  $ TZVALUE          : num  0.909 1.589 0.325 0.492 1.069 ...</span></span>
<span><span class="co">##  $ CONCINDEX1       : int  -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 ...</span></span>
<span><span class="co">##  $ CONCINDEX2       : int  0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##  $ SAMPLE1          : int  23 23 23 23 23 23 23 23 23 23 ...</span></span>
<span><span class="co">##  $ SAMPLE2          : int  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">##  $ SCORE            : int  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">##  $ COMBODRUGSEQ     : int  5587379 5587364 5587194 5587384 5587284 5587324 5587244 5587369 5587429 5587229 ...</span></span>
<span><span class="co">##  $ STUDY            : chr  "1307AC31" "1307AC31" "1307AC31" "1307AC31" ...</span></span>
<span><span class="co">##  $ TESTDATE         : chr  "07/08/2013" "07/08/2013" "07/08/2013" "07/08/2013" ...</span></span>
<span><span class="co">##  $ PLATE            : chr  "3577" "3432" "3548" "3606" ...</span></span>
<span><span class="co">##  $ SCREENER         : chr  "1A" "1A" "1A" "1A" ...</span></span>
<span><span class="co">##  $ PREFIX2          : chr  "" "" "" "" ...</span></span>
<span><span class="co">##  $ bio_rep          : int  1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##  $ tech_rep         : int  1 1 1 1 1 1 1 1 1 1 ...</span></span></code></pre>
<p>Examining the columns in our data, we see we have already defined the
both <code>bio_rep</code>s and <code>tech_rep</code>s during raw data in
preprocessing. We identified biological replicates as observations of
the sample treatment, dose and sample combination across multiple dates.
Technical replicates were the defined as any additional rows not mapped
to treatment, dose, sample and biological replicates.</p>
<p>For your own data in the wild, you will likely need to consult the
original publication and explore the data to differentiate these two
types of replicates and assign identifiers which make every observation
within an experiment unique.</p>
<p>For our second guess we will include these columns, mapping
biological replicates to samples and technical replicates to
treatments.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">groups2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    rowDataMap<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"treatment1dose"</span>,</span>
<span>        <span class="st">"treatment2dose"</span>, <span class="st">"tech_rep"</span><span class="op">)</span>,</span>
<span>    colDataMap<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    assayMap<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"treatment1dose"</span>,</span>
<span>        <span class="st">"treatment2dose"</span>,  <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span>, <span class="st">"tech_rep"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">guess2</span> <span class="op">&lt;-</span> <span class="fu">guessMapping</span><span class="op">(</span><span class="va">treDataMapper</span>, groups<span class="op">=</span><span class="va">groups2</span>, subset<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [CoreGx::guessMapping,LongTableDataMapper-method]</span></span>
<span><span class="co">##  Mapping for group rowDataMap: treatment1id, treatment2id, treatment1dose, treatment2dose, tech_rep</span></span></code></pre>
<pre><code><span><span class="co">## [CoreGx::guessMapping,LongTableDataMapper-method]</span></span>
<span><span class="co">##  Mapping for group colDataMap: sampleid, bio_rep</span></span></code></pre>
<pre><code><span><span class="co">## [CoreGx::guessMapping,LongTableDataMapper-method]</span></span>
<span><span class="co">##  Mapping for group assayMap: treatment1id, treatment2id, treatment1dose, treatment2dose, sampleid, bio_rep, tech_rep</span></span></code></pre>
<pre><code><span><span class="co">## $metadata</span></span>
<span><span class="co">## $metadata$id_columns</span></span>
<span><span class="co">## [1] NA</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $metadata$mapped_columns</span></span>
<span><span class="co">## character(0)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $rowDataMap</span></span>
<span><span class="co">## $rowDataMap$id_columns</span></span>
<span><span class="co">## [1] "treatment1id"   "treatment2id"   "treatment1dose" "treatment2dose"</span></span>
<span><span class="co">## [5] "tech_rep"      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $rowDataMap$mapped_columns</span></span>
<span><span class="co">## [1] "CONC1"      "CONC2"      "NSC2"       "CONCINDEX2"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $colDataMap</span></span>
<span><span class="co">## $colDataMap$id_columns</span></span>
<span><span class="co">## [1] "sampleid" "bio_rep" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $colDataMap$mapped_columns</span></span>
<span><span class="co">## [1] "CELLNAME" "CELLNBR"  "PANEL"    "PANELNBR"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $assayMap</span></span>
<span><span class="co">## $assayMap$id_columns</span></span>
<span><span class="co">## [1] "treatment1id"   "treatment2id"   "treatment1dose" "treatment2dose"</span></span>
<span><span class="co">## [5] "sampleid"       "bio_rep"        "tech_rep"      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $assayMap$mapped_columns</span></span>
<span><span class="co">##  [1] "viability"         "PERCENTGROWTH"     "PERCENTGROWTHNOTZ"</span></span>
<span><span class="co">##  [4] "EXPECTEDGROWTH"    "TESTVALUE"         "CONTROLVALUE"     </span></span>
<span><span class="co">##  [7] "TZVALUE"           "CONCINDEX1"        "SAMPLE1"          </span></span>
<span><span class="co">## [10] "SAMPLE2"           "SCORE"             "COMBODRUGSEQ"     </span></span>
<span><span class="co">## [13] "STUDY"             "TESTDATE"          "PLATE"            </span></span>
<span><span class="co">## [16] "SCREENER"          "PREFIX2"          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $unmapped</span></span>
<span><span class="co">## character(0)</span></span></code></pre>
<p>Since we no longer have any unmapped columns, we know that we have
found valid dimension identifiers for our data. It is possible that some
datasets will have multiple valid dimension definitions, so further
exploration may be required if the first solution doesn’t fit with your
domain knowledge of the experiment or with the design of the
<code>TRE</code>, which stores data more efficiently when only one
dimension is large.</p>
<p>Lets assign our mapped columns back to the <code>TREDataMapper</code>
so we can use it to construct a
<code>TreatmentResponseExperiment</code>.</p>
<p>For identified metadata, we need to assign a name to be used in the
<code>metadata</code> list of the createed <code>TRE</code>, other
dimension maps can be assigned directly to the slots of the
<code>TREDataMapper</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metadataMap</span><span class="op">(</span><span class="va">treDataMapper</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>experiment_metadata<span class="op">=</span><span class="va">guess2</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">mapped_columns</span><span class="op">)</span></span>
<span><span class="fu">rowDataMap</span><span class="op">(</span><span class="va">treDataMapper</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">guess2</span><span class="op">$</span><span class="va">rowDataMap</span></span>
<span><span class="fu">colDataMap</span><span class="op">(</span><span class="va">treDataMapper</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">guess2</span><span class="op">$</span><span class="va">colDataMap</span></span></code></pre></div>
<p>Unless we wish to have a single large assay in our <code>TRE</code>,
we should further subdivide our assay columns into sensible groups.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">guess2</span><span class="op">$</span><span class="va">assayMap</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $id_columns</span></span>
<span><span class="co">## [1] "treatment1id"   "treatment2id"   "treatment1dose" "treatment2dose"</span></span>
<span><span class="co">## [5] "sampleid"       "bio_rep"        "tech_rep"      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $mapped_columns</span></span>
<span><span class="co">##  [1] "viability"         "PERCENTGROWTH"     "PERCENTGROWTHNOTZ"</span></span>
<span><span class="co">##  [4] "EXPECTEDGROWTH"    "TESTVALUE"         "CONTROLVALUE"     </span></span>
<span><span class="co">##  [7] "TZVALUE"           "CONCINDEX1"        "SAMPLE1"          </span></span>
<span><span class="co">## [10] "SAMPLE2"           "SCORE"             "COMBODRUGSEQ"     </span></span>
<span><span class="co">## [13] "STUDY"             "TESTDATE"          "PLATE"            </span></span>
<span><span class="co">## [16] "SCREENER"          "PREFIX2"</span></span></code></pre>
<p>For the NCI-ALMANAC we have decided to separate our
treatment-response measurements into the “response” assay. Any
additional metadata which doesn’t map cleaning to rows or columns gets
assigned to “assay_metadata”. The publsihed synergy score will go in the
“profiles” assay, which generally stores metrics computed from the
“response” assay.</p>
<p>For our assay definitions we setup a nested list where each item
defines an assay in the <code>TRE</code>, with the “id_columns”
containing identifiers needed to uniquely map each row in that assay to
its associated treatments and samples. The “mapped_columns” item
specifies the definition of the assay table.</p>
<p>For this case, all of our assays have the same identifier columns.
However, this is not a requirement and it may be more efficient to test
further subgroupings with our identifier columns to prevent uncessarily
repeating values in our <code>TRE</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assaymap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    response<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        id_columns<span class="op">=</span><span class="va">guess2</span><span class="op">$</span><span class="va">assayMap</span><span class="op">$</span><span class="va">id_columns</span>,</span>
<span>        mapped_columns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"viability"</span>, <span class="st">"PERCENTGROWTH"</span>, <span class="st">"PERCENTGROWTHNOTZ"</span>,</span>
<span>            <span class="st">"EXPECTEDGROWTH"</span>, <span class="st">"TESTVALUE"</span>, <span class="st">"CONTROLVALUE"</span>, <span class="st">"TZVALUE"</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    profiles<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        id_columns<span class="op">=</span><span class="va">guess2</span><span class="op">$</span><span class="va">assayMap</span><span class="op">$</span><span class="va">id_columns</span>,</span>
<span>        mapped_columns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SCORE"</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    assay_metadata<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        id_columns<span class="op">=</span><span class="va">guess2</span><span class="op">$</span><span class="va">assayMap</span><span class="op">$</span><span class="va">id_columns</span>,</span>
<span>        mapped_columns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"COMBODRUGSEQ"</span>, <span class="st">"STUDY"</span>, <span class="st">"TESTDATE"</span>, <span class="st">"PLATE"</span>,</span>
<span>            <span class="st">"SCREENER"</span>, <span class="st">"PREFIX2"</span>, <span class="st">"CONCINDEX1"</span>, <span class="st">"CONCINDEX2"</span>, <span class="st">"SAMPLE1"</span>,</span>
<span>            <span class="st">"SAMPLE2"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">assayMap</span><span class="op">(</span><span class="va">treDataMapper</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">assaymap</span></span>
<span><span class="op">(</span><span class="va">treDataMapper</span><span class="op">)</span></span></code></pre></div>
<pre><code>##  [33m [1m [3m&lt;TREDataMapper&gt; 
##  [23m [22m [39m [33m [1mrawdata: [22m [39m dim(534179, 32)
##                              treatment1id treatment2id treatment1dose
##                                    &lt;char&gt;       &lt;char&gt;          &lt;num&gt;
##   1: 1,3-Bis(2-chloroethyl)-1-nitrosourea                        0.05
##   2: 1,3-Bis(2-chloroethyl)-1-nitrosourea                        0.05
##   3: 1,3-Bis(2-chloroethyl)-1-nitrosourea                        0.05
##   29 variables not shown: [treatment2dose &lt;num&gt;, sampleid &lt;char&gt;, CONC1
##     &lt;num&gt;, CONC2 &lt;num&gt;, NSC2 &lt;int&gt;, CELLNAME &lt;char&gt;, CELLNBR &lt;int&gt;,
##     PANEL &lt;char&gt;, PANELNBR &lt;int&gt;, viability &lt;num&gt;, ...] 
## 
 [33m [1mrowDataMap: [22m [39m
##    [32mrowIDs: [39m treatment1id, treatment2id, treatment1dose, treatment2dose, tech_rep
##    [32mrowMeta: [39m CONC1, CONC2, NSC2, CONCINDEX2
##  [33m [1mcolDataMap: [22m [39m
##    [32mcolIDs: [39m sampleid, bio_rep
##    [32mcolMeta: [39m CELLNAME, CELLNBR, PANEL, PANELNBR [33m [1m
## assayMap: [22m [39m
##    [31mresponse: [39m
##     keys: treatment1id, treatment2id, treatment1dose, treatment2dose, sampleid,
##       bio_rep, tech_rep
##     values: viability, PERCENTGROWTH, PERCENTGROWTHNOTZ, EXPECTEDGROWTH, TESTVALUE,
##       CONTROLVALUE, TZVALUE
##    [31mprofiles: [39m
##     keys: treatment1id, treatment2id, treatment1dose, treatment2dose, sampleid,
##       bio_rep, tech_rep
##     values: SCORE
##    [31massay_metadata: [39m
##     keys: treatment1id, treatment2id, treatment1dose, treatment2dose, sampleid,
##       bio_rep, tech_rep
##     values: COMBODRUGSEQ, STUDY, TESTDATE, PLATE, SCREENER, PREFIX2, CONCINDEX1,
##       CONCINDEX2, SAMPLE1, SAMPLE2 [33m [1m
## metadataMap: [22m [39m
##    [32mexperiment_metadata: [39m</code></pre>
<p>Now that we have assigned all the necessary metadata to the
<code>TREDataMapper</code>, creating the object is as simple as calling
<code>metaConstruct</code>, which combines together the rawdata and the
metadata into a constructor call for the
<code>TreatmentResponseExperiment</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">NCI_ALMANAC_tre</span> <span class="op">&lt;-</span> <span class="fu">metaConstruct</span><span class="op">(</span><span class="va">treDataMapper</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BBBB00; font-weight: bold; font-style: italic;">&lt;TreatmentResponseExperiment&gt; </span></span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold; font-style: italic;">## </span><span style="color: #BBBB00; font-weight: bold;">dim:  10080 62 </span></span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">## assays(3): </span><span style="color: #BB0000;">response profiles assay_metadata</span> </span></span>
<span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">rownames(10080): </span><span style="color: #00BB00;">1,3-Bis(2-chloroethyl)-1-nitrosourea::0.05:NA:1 1,3-Bis(2-chloroethyl)-1-nitrosourea::0.15:NA:1 ... Vorinostat:Vincristine:1:0.01:1 Vorinostat:Vincristine:1:0.1:1</span> </span></span>
<span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">rowData(9): </span><span style="color: #00BB00;">treatment1id treatment2id treatment1dose ... CONC2 NSC2 CONCINDEX2</span> </span></span>
<span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">colnames(62): </span><span style="color: #00BB00;">786-O:1 A-498:1 A-549:1 ... UACC-257:1 UACC-62:1 UO-31:1</span> </span></span>
<span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">colData(6): </span><span style="color: #00BB00;">sampleid bio_rep CELLNAME CELLNBR PANEL PANELNBR</span> </span></span>
<span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">metadata(1): </span><span style="color: #00BB00;">experiment_metadata</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="computing-over-a-treatmentresponseexperiment">Computing Over a <code>TreatmentResponseExperiment</code><a class="anchor" aria-label="anchor" href="#computing-over-a-treatmentresponseexperiment"></a>
</h4>
<p>So far we have seen how a <code>TRE</code> allows for flexible
definitions of treatments and samples to enable storage of dose-response
observations across a range of experimental designs. However, data
storage is not the sole point of the <code>TRE</code>. Once our data is
curated and stored safely inside our object we are ready to start
answering interesting questions about the experiment.</p>
<p>To facilitate computation over a <code>TRE</code>, we have included
the <code>aggregate</code> and <code>endoaggregate</code> methods which
allows arbitrary R functions to be applied over groups of identifiers
within an assay. For example, we may want to take averages for all
monotherapy viabilities over technical replicates. We can do so using
the <code>aggregate</code> method as follows:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">NCI_ALMANAC_tre</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">treatment2id</span> <span class="op">==</span> <span class="st">""</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span></span>
<span>    assay<span class="op">=</span><span class="st">"response"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">viability</span><span class="op">)</span>,</span>
<span>    by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment1dose"</span>, <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span></span>
<span>  <span class="va">mono_response_no_reps</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                treatment1id treatment1dose sampleid bio_rep</span></span>
<span><span class="co">##     1: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05    786-O       1</span></span>
<span><span class="co">##     2: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05    A-498       1</span></span>
<span><span class="co">##     3: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05    A-549       1</span></span>
<span><span class="co">##     4: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05     ACHN       1</span></span>
<span><span class="co">##     5: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05   BT-549       1</span></span>
<span><span class="co">##    ---                                                                     </span></span>
<span><span class="co">## 15181:                           Vorinostat          20.00    TK-10       1</span></span>
<span><span class="co">## 15182:                           Vorinostat          20.00  U-251MG       1</span></span>
<span><span class="co">## 15183:                           Vorinostat          20.00 UACC-257       1</span></span>
<span><span class="co">## 15184:                           Vorinostat          20.00  UACC-62       1</span></span>
<span><span class="co">## 15185:                           Vorinostat          20.00    UO-31       1</span></span>
<span><span class="co">##        mean_viability</span></span>
<span><span class="co">##     1:       98.89562</span></span>
<span><span class="co">##     2:       98.63179</span></span>
<span><span class="co">##     3:      100.91533</span></span>
<span><span class="co">##     4:      100.81652</span></span>
<span><span class="co">##     5:       95.67643</span></span>
<span><span class="co">##    ---               </span></span>
<span><span class="co">## 15181:       35.99182</span></span>
<span><span class="co">## 15182:       11.43731</span></span>
<span><span class="co">## 15183:       31.02885</span></span>
<span><span class="co">## 15184:       17.48634</span></span>
<span><span class="co">## 15185:       25.83149</span></span></code></pre>
<p>As you can see above, we select only our monotherapy response
observations using the <code>subset</code> method. Aggregate provides an
argument for the <code>assay</code> we want compute over, in this case
the “response” assay. We also specify how we want our groups defined
using the <code>by</code> argument. Finally, we can specify one or more
functions to aggregate columns with. In this case, we are simply taking
the mean of viability within groups defined by <code>by</code>. We have
excluded “tech_rep” for the by argument which means that our summary
will be computed over it.</p>
<p>The resulting table will contain the identifer columns specified to
<code>by</code>, as well as any columns which were summarized. By
default, for unnamed aggregations, <code>aggregate</code> tries to guess
the column name. We can see in the results it figured out that we are
taking the mean of the viability column and returned the results as
“mean_viability”.</p>
<p>You can also customize your resulting column names by naming the
aggregation arguments:</p>
<p>For example:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">NCI_ALMANAC_tre</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span></span>
<span>    subset<span class="op">=</span><span class="va">treatment2id</span> <span class="op">==</span> <span class="st">""</span>,</span>
<span>    assay<span class="op">=</span><span class="st">"response"</span>,</span>
<span>    my_viability<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">viability</span><span class="op">)</span>,</span>
<span>    my_dose<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">treatment1dose</span><span class="op">)</span>,</span>
<span>    by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment1dose"</span>, <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span></span>
<span>  <span class="va">mono_response_custom_names</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                treatment1id treatment1dose sampleid bio_rep</span></span>
<span><span class="co">##     1: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05    786-O       1</span></span>
<span><span class="co">##     2: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05    A-498       1</span></span>
<span><span class="co">##     3: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05    A-549       1</span></span>
<span><span class="co">##     4: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05     ACHN       1</span></span>
<span><span class="co">##     5: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05   BT-549       1</span></span>
<span><span class="co">##    ---                                                                     </span></span>
<span><span class="co">## 15181:                           Vorinostat          20.00    TK-10       1</span></span>
<span><span class="co">## 15182:                           Vorinostat          20.00  U-251MG       1</span></span>
<span><span class="co">## 15183:                           Vorinostat          20.00 UACC-257       1</span></span>
<span><span class="co">## 15184:                           Vorinostat          20.00  UACC-62       1</span></span>
<span><span class="co">## 15185:                           Vorinostat          20.00    UO-31       1</span></span>
<span><span class="co">##        my_viability my_dose</span></span>
<span><span class="co">##     1:     98.89562    0.05</span></span>
<span><span class="co">##     2:     98.63179    0.05</span></span>
<span><span class="co">##     3:    100.91533    0.05</span></span>
<span><span class="co">##     4:    100.81652    0.05</span></span>
<span><span class="co">##     5:     95.67643    0.05</span></span>
<span><span class="co">##    ---                     </span></span>
<span><span class="co">## 15181:     35.99182   20.00</span></span>
<span><span class="co">## 15182:     11.43731   20.00</span></span>
<span><span class="co">## 15183:     31.02885   20.00</span></span>
<span><span class="co">## 15184:     17.48634   20.00</span></span>
<span><span class="co">## 15185:     25.83149   20.00</span></span></code></pre>
<p>Here we can see that the viability column is named “my_viability” and
that our dose column named “my_dose”. We also use the
<code>subset</code> parameter, which was added to <code>aggregate</code>
for convenience to filter to relevant assay rows in a single function
call.</p>
<p>For more complex operations, such as optimizing parameters to a
non-linear function (as is done for fitting the Hill curve model to
dose-response data), we have provided the <code>nthread</code> argument
to parallelize computations when we have additional compute resources
available. In generaly, parallelized computations with aggregate will
use much more RAM than their serialized counterparts. Since there is
also overhead associated with dispatching parallel threads for a
computation, this feature should be used sparingly for only complex and
time consuming computations.</p>
<!-- We will see examples of parallelized aggregation in subsequent sections. -->
<p>The final feature of <code>aggregate</code> is allowing information
be shared between summarized columns to prevent the need for multiple
parallelized calls On our <code>TRE</code> object. For these cases, we
set the <code>enlist</code> parameter to <code>FALSE</code> and pass a
block of code wrapped in <code><a href="https://rdrr.io/r/base/Paren.html" class="external-link">{}</a></code>. The final statment in the
curly brackets should return a named list, where each list item will
become a column in the returned table.</p>
<p>The utility of this option will also become apparent in subsequent
sections.</p>
</div>
<div class="section level4">
<h4 id="combining-with-omics-data-into-a-pharmacoset">Combining with Omics Data into a <code>PharmacoSet</code><a class="anchor" aria-label="anchor" href="#combining-with-omics-data-into-a-pharmacoset"></a>
</h4>
<p>The NCI-ALMANAC study did not conduct its own set of molecular
profiling experiments, since the NCI60 already profiled all cell-lines
available in the NCI-ALMANAC. We have already curated a
<code>PharmacoSet</code> for <code>NCI60</code>, so adding our molecular
data to the NCI-ALMANAC <code>PharmacoSet</code> requires no manual
curation. We must simply subset the <code>NCI60</code> dataset to those
cell-lines present in NCI-ALMANAC.</p>
<p>Lets examine the included molecualr data:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">NCI60_molecular_data</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">NCI60_molecular_data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## A MultiAssayExperiment object of 4 listed</span></span>
<span><span class="co">##  experiments with user-defined names and respective classes.</span></span>
<span><span class="co">##  Containing an ExperimentList class object of length 4:</span></span>
<span><span class="co">##  [1] rna: RangedSummarizedExperiment with 54609 rows and 59 columns</span></span>
<span><span class="co">##  [2] mirna: RangedSummarizedExperiment with 417 rows and 59 columns</span></span>
<span><span class="co">##  [3] rnaseq.comp: RangedSummarizedExperiment with 23808 rows and 59 columns</span></span>
<span><span class="co">##  [4] rnaseq.iso: RangedSummarizedExperiment with 23808 rows and 59 columns</span></span>
<span><span class="co">## Functionality:</span></span>
<span><span class="co">##  experiments() - obtain the ExperimentList instance</span></span>
<span><span class="co">##  colData() - the primary/phenotype DataFrame</span></span>
<span><span class="co">##  sampleMap() - the sample coordination DataFrame</span></span>
<span><span class="co">##  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span></span>
<span><span class="co">##  *Format() - convert into a long or wide DataFrame</span></span>
<span><span class="co">##  assays() - convert ExperimentList to a SimpleList of matrices</span></span>
<span><span class="co">##  exportClass() - save data to flat files</span></span></code></pre>
<p>We can see the molecular data is available as
<code>MultiAssayExperiment</code> containing four
<code>RangedSummarizedExperiment</code> objects. The molecular data
types are:</p>
<ol style="list-style-type: decimal">
<li>“rna” represents RNA expression microarray data.</li>
<li>“mirna” represents microRNA expression data.</li>
<li>“rnaseq.comp” contains RNA sequencing data.</li>
<li>“rnaseq.iso” is RNA sequencing data specifically targeting
transcript isoforms.</li>
</ol>
<p>Once we have our <code>MultiAssayExperiment</code> and
<code>TreatmentResponseExperiment</code> objects ready, we next have to
extract and curate sample and treatment metadata tables which store a
summary of metadata for available samples and treatments in our
<code>PharmacoSet</code> object.</p>
<p>For brevity, we have included preassembled annotation files for
NCI-ALMANAC since manual curation is beyong the scope of this workshop.
When creating your own <code>PharmacoSet</code>, we recommend assembling
a set of standardized annotations from sources like PubChem or ChEMBL
for treatments and Cellosaurus for cancer cell-lines and using these to
standardize identifiers in your data as well as adding links to
additional information in external databases.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">NCI_ALMANAC_sample_metadata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">NCI_ALMANAC_sample_metadata</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="5%">
<col width="5%">
<col width="5%">
<col width="32%">
<col width="9%">
<col width="13%">
<col width="17%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">sampleid</th>
<th align="left">tissueid</th>
<th align="left">cellosaurus_disease</th>
<th align="left">cellosaurus_id</th>
<th align="left">NCI_ALMANAC.sampleid</th>
<th align="left">NCI_ALMANAC.disease</th>
<th align="left">pharmacodb_cid</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">786-O</td>
<td align="left">786-O</td>
<td align="left">Kidney</td>
<td align="left">Renal cell carcinoma</td>
<td align="left">CVCL_1051</td>
<td align="left">786-0</td>
<td align="left">Renal Cancer</td>
<td align="left">7860_28_2019</td>
</tr>
<tr class="even">
<td align="left">A-498</td>
<td align="left">A-498</td>
<td align="left">Kidney</td>
<td align="left">Renal cell carcinoma</td>
<td align="left">CVCL_1056</td>
<td align="left">A498</td>
<td align="left">Renal Cancer</td>
<td align="left">A498_47_2019</td>
</tr>
<tr class="odd">
<td align="left">A-549</td>
<td align="left">A-549</td>
<td align="left">Lung</td>
<td align="left">Lung adenocarcinoma</td>
<td align="left">CVCL_0023</td>
<td align="left">A549/ATCC</td>
<td align="left">Non-Small Cell Lung Cancer</td>
<td align="left">A549_48_2019</td>
</tr>
<tr class="even">
<td align="left">ACHN</td>
<td align="left">ACHN</td>
<td align="left">Kidney</td>
<td align="left">Papillary renal cell carcinoma</td>
<td align="left">CVCL_1067</td>
<td align="left">ACHN</td>
<td align="left">Renal Cancer</td>
<td align="left">ACHN_54_2019</td>
</tr>
<tr class="odd">
<td align="left">BT-549</td>
<td align="left">BT-549</td>
<td align="left">Breast</td>
<td align="left">Invasive ductal carcinoma, not otherwise specified</td>
<td align="left">CVCL_1092</td>
<td align="left">BT-549</td>
<td align="left">Breast Cancer</td>
<td align="left">BT549_111_2019</td>
</tr>
<tr class="even">
<td align="left">CCRF-CEM</td>
<td align="left">CCRF-CEM</td>
<td align="left">Lymphoid</td>
<td align="left">Childhood T acute lymphoblastic leukemia</td>
<td align="left">CVCL_0207</td>
<td align="left">CCRF-CEM</td>
<td align="left">Leukemia</td>
<td align="left">CCRFCEM_190_2019</td>
</tr>
</tbody>
</table>
<p>The above table contains the sample metadata which will comprise the
<code>@sample</code> slot of our <code>PharmacoSet</code>. It contains a
set of standardized identifiers including <code>cellid</code> and
<code>tissueid</code>, as well as cell-line identifiers from the
NCI-ALMANAC publication and some additional metadata we have added.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">NCI_ALMANAC_treatment_metadata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">NCI_ALMANAC_treatment_metadata</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="14%">
<col width="14%">
<col width="9%">
<col width="3%">
<col width="46%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">treatmentid</th>
<th align="left">NCI_ALMANAC.treatmentid</th>
<th align="right">cid</th>
<th align="left">smiles</th>
<th align="left">inchikey</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Pralatrexate</td>
<td align="left">Pralatrexate</td>
<td align="left">pralatrexate</td>
<td align="right">148121</td>
<td align="left">C#CCC(CC1=CN=C2C(=N1)C(=NC(=N2)N)N)C3=CC=C(C=C3)C(=O)NC(CCC(=O)O)C(=O)O</td>
<td align="left">OGSBUKJUDHAQEA-WMCAAGNKSA-N</td>
</tr>
<tr class="even">
<td align="left">1,3-Bis(2-chloroethyl)-1-nitrosourea</td>
<td align="left">1,3-Bis(2-chloroethyl)-1-nitrosourea</td>
<td align="left">Carmustine</td>
<td align="right">2578</td>
<td align="left">C(CCl)NC(=O)N(CCCl)N=O</td>
<td align="left">DLGOEMSEDOSKAD-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td align="left">Ixabepilone</td>
<td align="left">Ixabepilone</td>
<td align="left">Ixabepilone</td>
<td align="right">6445540</td>
<td align="left">CC1CCCC2(C(O2)CC(NC(=O)CC(C(C(=O)C(C1O)C)(C)C)O)C(=CC3=CSC(=N3)C)C)C</td>
<td align="left">FABUFPQFXZVHFB-PVYNADRNSA-N</td>
</tr>
<tr class="even">
<td align="left">Clofarabine</td>
<td align="left">Clofarabine</td>
<td align="left">Clofarabine</td>
<td align="right">119182</td>
<td align="left">Nc1nc(Cl)nc2n(cnc12)C3OC(CO)C(O)C3F</td>
<td align="left">WDDPHFBMKLOVOX-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td align="left">Dasatinib</td>
<td align="left">Dasatinib</td>
<td align="left">Dasatinib</td>
<td align="right">3062316</td>
<td align="left">CC1=C(C(=CC=C1)Cl)NC(=O)C2=CN=C(S2)NC3=NC(=NC(=C3)N4CCN(CC4)CCO)C</td>
<td align="left">ZBNZXTGUTAYRHI-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td align="left">Vincristine</td>
<td align="left">Vincristine</td>
<td align="left">Vincristine sulfate</td>
<td align="right">5978</td>
<td align="left">CCC1(O)CC2CN(C1)CCc3c([nH]c4ccccc34)C(C2)(C(=O)OC)c5cc6c(cc5OC)N(C=O)C7C86CCN9CC=CC(CC)(C98)C(C(=O)OC)C7(O)C(=O)OC</td>
<td align="left"></td>
</tr>
</tbody>
</table>
<p>Similar to the sample table, this one contains treatment metadata
which will be assigned to the <code>@treatment</code> slot of our
<code>PharmacoSet</code>. It also contains a set of standardized
identifiers including <code>treatmentid</code>, structural chemical
identifiers, links to external databases as well as the drug identifiers
from the NCI-ALMANAC publication.</p>
<p>For the curation list, we extract our mappings from our standardized
identifiers to dataset specific identifiers and store those separately
to ensure these associations are not lost should the treatment or sample
metadata tables be modified.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">curation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  sample<span class="op">=</span><span class="va">NCI_ALMANAC_sample_metadata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sampleid"</span>, <span class="st">"NCI_ALMANAC.sampleid"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>  treatment<span class="op">=</span><span class="va">NCI_ALMANAC_treatment_metadata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatmentid"</span>, <span class="st">"NCI_ALMANAC.treatmentid"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>  tissue<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="va">NCI_ALMANAC_sample_metadata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tissueid"</span>, <span class="st">"NCI_ALMANAC.disease"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    rownames<span class="op">=</span><span class="va">NCI_ALMANAC_sample_metadata</span><span class="op">$</span><span class="va">sampleid</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>With all of our pieces in place, we can finally construct a drug
combination <code>PharmacoSet</code> which can be used in subsequent
analysis. Along with the recent changes in the structure and
nomenclature of a <code>PharmacoSet</code> comes the introduction of the
<code>PharmacoSet2</code> contructor. This version enforces new
requirements for the object such as a <code>MultiAssayExperiment</code>
for the molecular profiles slot and a
<code>TreatmentResponseExperiment</code> for treatment response slot.
The original constructor is still included in the package for backwards
compatibility but will be retired after we finish updating all our
existing datasets on ORCESTRA.ca.</p>
<p>If you have an old style <code>PharmacoSet</code> but want to use the
new package features, simply call <code>updateObject</code> to return
the dataset with an updated class structure. This method will be
available in all future releases to ensure datasets can be updated and
reused regardless of their age.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">NCI_ALMANAC_2017</span> <span class="op">&lt;-</span> <span class="fu">PharmacoSet2</span><span class="op">(</span></span>
<span>  name<span class="op">=</span><span class="st">"NCI_ALMANAC_2017"</span>,</span>
<span>  molecularProfiles<span class="op">=</span><span class="va">NCI60_molecular_data</span>,</span>
<span>  treatmentResponse<span class="op">=</span><span class="va">NCI_ALMANAC_tre</span>,</span>
<span>  treatment<span class="op">=</span><span class="va">NCI_ALMANAC_treatment_metadata</span>,</span>
<span>  sample<span class="op">=</span><span class="va">NCI_ALMANAC_sample_metadata</span>,</span>
<span>  curation<span class="op">=</span><span class="va">curation</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## &lt;PharmacoSet&gt;</span></span>
<span><span class="co">## Name:  NCI_ALMANAC_2017 </span></span>
<span><span class="co">## Date Created:  Thu Jul 28 17:24:00 2022 </span></span>
<span><span class="co">## Number of samples:  60 </span></span>
<span><span class="co">## Molecular profiles:</span></span>
<span><span class="co">##    A MultiAssayExperiment object of 4 listed</span></span>
<span><span class="co">##    experiments with user-defined names and respective classes.</span></span>
<span><span class="co">##    Containing an ExperimentList class object of length 4:</span></span>
<span><span class="co">##    [1] rna: RangedSummarizedExperiment with 54609 rows and 59 columns</span></span>
<span><span class="co">##    [2] mirna: RangedSummarizedExperiment with 417 rows and 59 columns</span></span>
<span><span class="co">##    [3] rnaseq.comp: RangedSummarizedExperiment with 23808 rows and 59 columns</span></span>
<span><span class="co">##    [4] rnaseq.iso: RangedSummarizedExperiment with 23808 rows and 59 columns </span></span>
<span><span class="co">## Treatment response:</span></span>
<span><span class="co">##    <span style="color: #BBBB00; font-weight: bold; font-style: italic;">&lt;TreatmentResponseExperiment&gt; </span></span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold; font-style: italic;">##   </span><span style="color: #BBBB00; font-weight: bold;">dim:  10080 62 </span></span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">##   assays(3): </span><span style="color: #BB0000;">response profiles assay_metadata</span> </span></span>
<span><span class="co">##   <span style="color: #BBBB00; font-weight: bold;">rownames(10080): </span><span style="color: #00BB00;">1,3-Bis(2-chloroethyl)-1-nitrosourea::0.05:NA:1 1,3-Bis(2-chloroethyl)-1-nitrosourea::0.15:NA:1 ... Vorinostat:Vincristine:1:0.01:1 Vorinostat:Vincristine:1:0.1:1</span> </span></span>
<span><span class="co">##   <span style="color: #BBBB00; font-weight: bold;">rowData(9): </span><span style="color: #00BB00;">treatment1id treatment2id treatment1dose ... CONC2 NSC2 CONCINDEX2</span> </span></span>
<span><span class="co">##   <span style="color: #BBBB00; font-weight: bold;">colnames(62): </span><span style="color: #00BB00;">786-O:1 A-498:1 A-549:1 ... UACC-257:1 UACC-62:1 UO-31:1</span> </span></span>
<span><span class="co">##   <span style="color: #BBBB00; font-weight: bold;">colData(6): </span><span style="color: #00BB00;">sampleid bio_rep CELLNAME CELLNBR PANEL PANELNBR</span> </span></span>
<span><span class="co">##   <span style="color: #BBBB00; font-weight: bold;">metadata(1): </span><span style="color: #00BB00;">experiment_metadata</span></span></span></code></pre>
<p>Finally we see our finished drug combination <code>PharmacoSet</code>
and are ready to use it to answer interesting research questions and
discover the relationship between molecular phenotypes and treatment
response!</p>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="dose-response-models-and-drug-sensitivity-measures">Dose Response Models and Drug Sensitivity Measures<a class="anchor" aria-label="anchor" href="#dose-response-models-and-drug-sensitivity-measures"></a>
</h2>
<p>Understanding monotherapy dose-viability data is an important step in
assessing drug combination response. Drug synergy is assessed by
comparing the observed combination response to what is expected from the
response of cells to each drug separately. While different models for
“expected” combination effect will be discussed in the next section, we
first review how monotherapy data are analyzed.</p>
<p>Monotherapy dose-response experiments are usually analyzed by first
fitting a dose-response or dose-viability model to the observed
measurements. The most common model used in the literature is a form of
sigmoid function, called the Hill Curve. The Hill Curve has its basis in
the Michaelis-Menten chemical kinetics model, while adding an extra
parameter that is traditionally interpreted to account for
co-operativity of binding. To better match real observed data, the model
can also be extended to include parameters controlling the maximum and
minimum observed viability.</p>
<p><em>NB: Sometimes, dose-viability and dose-response curves are used
interchangebly in the literature. Precisely, dose-response curves for a
cell kiling screen would plot % cell death against dose, leading to an
increasing sigmoid curve.</em></p>
<p>In <em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>,
we use the 3 Parameter Hill Curve function as our model of drug response
in cancer cell lines:</p>
<p><span class="math display">\[y = E_\infty + \frac{1-E_\infty}{1 +
(\frac{x}{EC50})^{HS}}\]</span></p>
<p>This is a log-logistic model, meaning that it takes the form of a
logistic curve when x is on the log scale. The three parameters are:
<span class="math inline">\(E_\infty\)</span>, which is the maximal
inhibition predicted at infinite concentration of the drug (the right
asymptote); <span class="math inline">\(EC50\)</span>, which is the
inflection point of the logit, where the inhibition reaches 50% of the
maximum; <span class="math inline">\(HS\)</span>, the Hill Slope, which
is thought to be a measure of the cooperativity of binding. This
parameter controls the steepness of the logit, and is interpreted based
on its relation to 1. Values larger than 1 imply positive cooperativity
in inhibition/binding of the target for the small molecule ligands,
which values less than 1 imply negative cooperativity/antagonism.</p>
<div class="figure">
<img src="hill_slope.png" alt=""><p class="caption">Diagram of a logLogistic Hill Slope</p>
</div>
<p>PharmacoGx provides functions to both fit and visualize
dose-viability curves. Plotting dose-viability data is easy using the
<code>drugDoseResponseCurve</code> function:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">concentrations</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span><span class="op">^</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">8</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1</span></span>
<span><span class="va">viabilities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">5</span>, <span class="fl">15</span>, <span class="fl">23.8</span>, <span class="fl">55.2</span>, <span class="fl">85.9</span>, <span class="fl">96.9</span>, <span class="fl">98.4</span>, <span class="fl">99.2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">drugDoseResponseCurve</span><span class="op">(</span></span>
<span>  concentrations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Exp 1"</span> <span class="op">=</span> <span class="va">concentrations</span><span class="op">)</span>,</span>
<span>  viabilities <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Exp 1"</span> <span class="op">=</span> <span class="va">viabilities</span><span class="op">)</span>,</span>
<span>  plot.type <span class="op">=</span> <span class="st">"Both"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="PGxWorkshop_files/figure-html/ddrc_own-1.png" width="700"></p>
<p>Similarly, fitting Log-Logistic models to data is simple using the
<code>logLogisticRegression</code> function in
<code>PharmacoGx</code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">concentrations</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span><span class="op">^</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">8</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1</span></span>
<span><span class="va">viabilities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">5</span>, <span class="fl">15</span>, <span class="fl">23.8</span>, <span class="fl">55.2</span>, <span class="fl">85.9</span>, <span class="fl">96.9</span>, <span class="fl">98.4</span>, <span class="fl">99.2</span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu">logLogisticRegression</span><span class="op">(</span>conc <span class="op">=</span> <span class="va">concentrations</span>, viability <span class="op">=</span> <span class="va">viabilities</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $HS</span></span>
<span><span class="co">## [1] 2.053572</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $E_inf</span></span>
<span><span class="co">## [1] 3.284808</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $EC50</span></span>
<span><span class="co">## [1] 0.06902647</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attr(,"Rsquare")</span></span>
<span><span class="co">## [1] 0.9965584</span></span></code></pre>
<!-- `(` always prints its contents, even for assignments. It's cleaner than printing after IMO. -->
<p>Note that sometimes, it is may be more appropriate to use either a 2
parameter model (excluding the <span class="math inline">\(E_\infty\)</span> parameter), or a 4 parameter
model, including a parameter for the left-assymptote, that is, the
measurement when no drug is added. <code>PharmacoGx</code> is opiniated
in using the 3 parameter model, as the logical value for the viability
as a percentage of untreated cells when no drug is added is 100%.
However, the <em><a href="https://CRAN.R-project.org/package=drc" class="external-link">drc</a></em> CRAN package
provides a larger variety of dose-response models, should it be
appropriate in your application.</p>
<p>We will soon demonstrate how to apply the
<code>logLogisticRegression</code> to dose-viability data stored in a
<code>TreatmentResponseExperiment</code>, but first we will talk about
how to summarize dose-response curves for downstream analysis.</p>
<div class="section level4">
<h4 id="computing-summary-measures-for-ddrcs">Computing Summary Measures for DDRCs<a class="anchor" aria-label="anchor" href="#computing-summary-measures-for-ddrcs"></a>
</h4>
<p>Often with <em>in vitro</em> Pharmacogenomics data, we want to
compare the drug sensitivity of a cell line to some omic features. For
this, we want to summarize the drug dose response curve into a single
number representing the sensitivity of the cell line.</p>
<div class="figure">
<img src="hillSlopeSummary.png" alt=""><p class="caption">Some commonly used summary metrics for a
dose-response curve</p>
</div>
<p>The IC50 and Area Above the Curve are two convenient metrics for
quantifying the observed drug sensitivity. If you noticed above, the
<code>drugDoseResponseCurve</code> function computes them by default. In
PharmacoGx, they can be computed manually as follows:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">concentrations</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span><span class="op">^</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">8</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1</span></span>
<span><span class="va">viabilities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">5</span>, <span class="fl">15</span>, <span class="fl">23.8</span>, <span class="fl">55.2</span>, <span class="fl">85.9</span>, <span class="fl">96.9</span>, <span class="fl">98.4</span>, <span class="fl">99.2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">PharmacoGx</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PharmacoGx/man/computeAUC.html" class="external-link">computeAUC</a></span><span class="op">(</span>concentration <span class="op">=</span> <span class="va">concentrations</span>, viability <span class="op">=</span> <span class="va">viabilities</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 46.63697</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">PharmacoGx</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PharmacoGx/man/computeICn.html" class="external-link">computeIC50</a></span><span class="op">(</span>concentration <span class="op">=</span> <span class="va">concentrations</span>, viability <span class="op">=</span> <span class="va">viabilities</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.0713488</span></span></code></pre>
<p>If measurements are passed into these function, the Log Logistic
model is fit to compute the IC50 and AAC. Clearly, if multiple metrics
are being computed, redoing the fit is inefficent. Therefore, these
functions can also take parameters fit by
<code>logLogisticRegression</code>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">concentrations</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span><span class="op">^</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">8</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1</span></span>
<span><span class="va">viabilities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">5</span>, <span class="fl">15</span>, <span class="fl">23.8</span>, <span class="fl">55.2</span>, <span class="fl">85.9</span>, <span class="fl">96.9</span>, <span class="fl">98.4</span>, <span class="fl">99.2</span><span class="op">)</span></span>
<span><span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu">logLogisticRegression</span><span class="op">(</span>conc <span class="op">=</span> <span class="va">concentrations</span>, viability <span class="op">=</span> <span class="va">viabilities</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">PharmacoGx</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PharmacoGx/man/computeAUC.html" class="external-link">computeAUC</a></span><span class="op">(</span>concentration <span class="op">=</span> <span class="va">concentrations</span>, Hill_fit <span class="op">=</span> <span class="va">pars</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 46.63697</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">PharmacoGx</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PharmacoGx/man/computeICn.html" class="external-link">computeIC50</a></span><span class="op">(</span>concentration <span class="op">=</span> <span class="va">concentrations</span>, Hill_fit <span class="op">=</span> <span class="va">pars</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.0713488</span></span></code></pre>
<p>As processsing monotherapy data is often a step in preparing for
assessing synergy from combination experiments, we will demonstrate how
to use these functions together with a TreatmentResponseExperiment
during our preprocessing for synergy calculation.</p>
</div>
</div>
<div class="section level2">
<h2 id="drug-combination-synergy-models">Drug Combination Synergy Models<a class="anchor" aria-label="anchor" href="#drug-combination-synergy-models"></a>
</h2>
<p>When assessing drug combinations, the question of interest is usually
not to quantify drug response to the therapy, but to identify
combinations that give more or less response to therapy than expected
from the monotherapy regiments, that is, to look for synergy or
antagonism. Alternatively, for a paritcular drug combination, we may be
interested to identify which biological models show synergy or
antagonism, and understand why that could be the case on a molecular
level.</p>
<p>For these tasks, it is important to define what we expect the result
of combining two drugs at a particular combination to be without the
presence of any synergy or antagonism between them. Usually, when
searching for synergy, will have access to experiments testing each drug
in monotherapy, as well as the drugs in combination. For our discussion,
we we restrict ourselves to the two-drug combination case, however, the
principles we discuss can be extended to handle multiple drugs.</p>
<p>Unfortunately, there is no consensus of what the theoretical expected
value of response for combining two drugs is, given the response that
each drug elicits in monotherapy. However, there are a few models widely
applied in the literature, each with different biological assumptions
behind them.</p>
<p>To introduce these models, we will define some mathematical notation,
which we will use both throughout our exposition and the example
computations. Let <span class="math inline">\(x_{1}\)</span>, and <span class="math inline">\(x_{2}\)</span> correspond to the corresponding
drug 1 and drug 2 doses at which we wish to predict the expected
combination response. Let <span class="math inline">\(v_{1}\)</span> and
<span class="math inline">\(v_{2}\)</span> be the viabilities that are
observed in monotherapy for these drugs, and <span class="math inline">\(v_{c}\)</span> be the viability observed in
combination. Furthermore, let <span class="math inline">\(r_{i} = 1 -
v_{i}\)</span> be the response value corresponding to each measured
viability. We will use both response and viability measurements in
explaining the models, as some are more natural from one or the other
perspective. However, it is easy to convert between these variables
using the relationship above.</p>
<p>When we look for drug synergy, we are looking for cases where <span class="math inline">\(v_c &lt; v_{model}\)</span>, or <span class="math inline">\(r_c &gt; r_{model}\)</span>. The degree of synergy
is usually measured either on the additive or multplicative scale, that
is, looking at either <span class="math inline">\(v_c -
v_{model}\)</span> or <span class="math inline">\(v_c /
v_{model}\)</span>. However, some models, such as the ZIP, also
perscribe a certain perspective or scoring method to assess synergy.</p>
<p>This vignette covers 4 different models for predicting what the
effect of drug combination treatment should be.</p>
<p><em>In the live workshop we focus only on the first 3
models.</em></p>
<ol style="list-style-type: decimal">
<li>Highest Single Agent</li>
<li>Bliss Independence</li>
<li>Loewe Additivity</li>
<li>Zero Interaction Potency (ZIP)</li>
</ol>
<p>As we introduce each method, we will go through the theoretical
motivation behind it, the calculation, and then show how to compute the
expected viability/response for each model using the
TreatmentResponseExperiment API.</p>
<div class="section level4">
<h4 id="preprocessing-monotherapy-data-for-synergy-estimation">Preprocessing Monotherapy Data for Synergy Estimation<a class="anchor" aria-label="anchor" href="#preprocessing-monotherapy-data-for-synergy-estimation"></a>
</h4>
<p>Prior to computing expected viabilities from combination experiments,
the monotherapy data may need to be preprocessed. In our case, the
dataset we are using has replicate experiments for the monotherapy
curves, as seen below:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">treatmentResponse</span><span class="op">(</span><span class="va">NCI_ALMANAC_2017</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">treatment2id</span> <span class="op">==</span> <span class="st">""</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span></span>
<span>      assay<span class="op">=</span><span class="st">"response"</span>,</span>
<span>      N<span class="op">=</span><span class="va">.N</span>,</span>
<span>      by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment1dose"</span>, <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span>expr<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">N</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="PGxWorkshop_files/figure-html/plotting_monotherapy_replicates-1.png" width="700"></p>
<p>The majority of these were replicates across plates, most probably so
that monotherapy results could be plate matched to combination results.
Therefore, we can make two choices for how to handle these replicates,
either average across plates, or to match analysis by plate whenever
possible. Ideally, we would recommend the latter approach, however, it
is computationally more expensive (as we must fit monotherapy curves
per-plate). In interest of computation time, we average replicates
across all experiments for our demonstration.</p>
<p>Here, we use the <code>endoaggregate</code> TRE method while setting
the <code>by</code> argument to compute an aggregation over each unique
combination of <code>treatment1id</code>, <code>treatment1dose</code>,
<code>sampleid</code> and <code>bio_rep</code> (summarizing over
technical replicates).</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">NCI_TRE</span> <span class="op">&lt;-</span> <span class="fu">treatmentResponse</span><span class="op">(</span><span class="va">NCI_ALMANAC_2017</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># -- Extract and summarize monotherapy data, defined by an absent treatment2id</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">endoaggregate</span><span class="op">(</span></span>
<span>    assay<span class="op">=</span><span class="st">"response"</span>,</span>
<span>    target<span class="op">=</span><span class="st">"mono_response"</span>,</span>
<span>    subset <span class="op">=</span> <span class="va">treatment2id</span> <span class="op">==</span> <span class="st">""</span>,</span>
<span>    viability<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">viability</span><span class="op">)</span>,</span>
<span>    by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment1dose"</span>, <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span></span>
<span>  <span class="va">NCI_TRE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">NCI_TRE</span><span class="op">$</span><span class="va">mono_response</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            treatment1id treatment1dose sampleid bio_rep</span></span>
<span><span class="co">## 1: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05    786-O       1</span></span>
<span><span class="co">## 2: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05    A-498       1</span></span>
<span><span class="co">## 3: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05    A-549       1</span></span>
<span><span class="co">## 4: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05     ACHN       1</span></span>
<span><span class="co">## 5: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05   BT-549       1</span></span>
<span><span class="co">## 6: 1,3-Bis(2-chloroethyl)-1-nitrosourea           0.05 CCRF-CEM       1</span></span>
<span><span class="co">##    viability</span></span>
<span><span class="co">## 1:  98.89562</span></span>
<span><span class="co">## 2:  98.63179</span></span>
<span><span class="co">## 3: 100.91533</span></span>
<span><span class="co">## 4: 100.81652</span></span>
<span><span class="co">## 5:  95.67643</span></span>
<span><span class="co">## 6: 102.38559</span></span></code></pre>
<p>Additionally, computing the Loewe and ZIP scores requires knowing the
Hill Fit parameters for the monotherapy experiments. This is also a
required step whenever it is necessary to interpolate between measured
monotherapy doses to predict for the measured combination viability
doses.</p>
<p>We can fit the curves using <code>logLogisticRegression</code>, and
save the monotherapy fits back to the
TreatmentResponseExperimentClassDiagram using the same endoaggregate
pattern as above. Since we are already calculating the Hill Curve fits
for these experiments, we will also calculate the AAC and IC50
values.</p>
<p>Note that by default, <code>logLogisticRegression</code> returns the
<code>E_inf</code> parameter on a percentage scale when the input data
is on a percentage scale. We convert to a fraction scale because that
makes the subsequent calculations simpler.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># -- Fit the Hill curve model to monotherapy drugs and compute associated metrics</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">endoaggregate</span><span class="op">(</span></span>
<span>    <span class="op">{</span></span>
<span>      <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">PharmacoGx</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PharmacoGx/man/logLogisticRegression.html" class="external-link">logLogisticRegression</a></span><span class="op">(</span><span class="va">treatment1dose</span>, <span class="va">viability</span><span class="op">)</span></span>
<span>      <span class="va">ic50</span> <span class="op">&lt;-</span> <span class="fu">PharmacoGx</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PharmacoGx/man/computeICn.html" class="external-link">computeIC50</a></span><span class="op">(</span><span class="va">treatment1dose</span>, Hill_fit <span class="op">=</span> <span class="va">fit</span><span class="op">)</span></span>
<span>      <span class="va">aac</span> <span class="op">&lt;-</span> <span class="fu">PharmacoGx</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PharmacoGx/man/computeAUC.html" class="external-link">computeAUC</a></span><span class="op">(</span><span class="va">treatment1dose</span>, Hill_fit <span class="op">=</span> <span class="va">fit</span>, area.type <span class="op">=</span> <span class="st">"Fitted"</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        HS <span class="op">=</span> <span class="va">fit</span><span class="op">[[</span><span class="st">"HS"</span><span class="op">]</span><span class="op">]</span>, E_inf <span class="op">=</span> <span class="va">fit</span><span class="op">[[</span><span class="st">"E_inf"</span><span class="op">]</span><span class="op">]</span><span class="op">/</span><span class="fl">100</span>, EC50 <span class="op">=</span> <span class="va">fit</span><span class="op">[[</span><span class="st">"EC50"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>        Rsq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        aac <span class="op">=</span> <span class="va">aac</span>,</span>
<span>        ic50 <span class="op">=</span> <span class="va">ic50</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">}</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"mono_response"</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    enlist <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    target <span class="op">=</span> <span class="st">"mono_profiles"</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span> <span class="va">NCI_TRE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">NCI_TRE</span><span class="op">$</span><span class="va">mono_profiles</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            treatment1id sampleid bio_rep HS      E_inf EC50</span></span>
<span><span class="co">## 1: 1,3-Bis(2-chloroethyl)-1-nitrosourea    786-O       1  1 0.08977556   50</span></span>
<span><span class="co">## 2: 1,3-Bis(2-chloroethyl)-1-nitrosourea    A-498       1  1 0.56096579  500</span></span>
<span><span class="co">## 3: 1,3-Bis(2-chloroethyl)-1-nitrosourea    A-549       1  1 0.15732265   50</span></span>
<span><span class="co">## 4: 1,3-Bis(2-chloroethyl)-1-nitrosourea     ACHN       1  1 0.01729107   50</span></span>
<span><span class="co">## 5: 1,3-Bis(2-chloroethyl)-1-nitrosourea   BT-549       1  1 0.05578801   20</span></span>
<span><span class="co">## 6: 1,3-Bis(2-chloroethyl)-1-nitrosourea CCRF-CEM       1  1 0.12852970   50</span></span>
<span><span class="co">##          Rsq       aac     ic50</span></span>
<span><span class="co">## 1: 0.9425047 23.687649 60.94225</span></span>
<span><span class="co">## 2: 0.7388963  3.303585       NA</span></span>
<span><span class="co">## 3: 0.8796848 21.929806 72.95492</span></span>
<span><span class="co">## 4: 0.8303083 25.573983 51.79104</span></span>
<span><span class="co">## 5: 0.4578065 33.375273 22.51177</span></span>
<span><span class="co">## 6: 0.9603414 22.679113 67.30013</span></span></code></pre>
<p>We will now extract the combination data into its own assay in the
LongTable, and append the monotherapy Hill fits using the
<code>mergeAssays</code> function, so that we can later use the fitted
data from above to calculate expected combination viability. We also
filter our downstream analysis only to those experiments where the
monotherapy fits were good (R squared better than 0.5).</p>
<p>Notice that we need to do two merges, one for appending the
monotherapy Hill fits for the first compound treatment, and one for
appending the monotherapy Hill fits for the second compound treatment.
Fortunately, the <code>suffixes</code> parameter will allow us to
differentiate between the two once they are merged to the same
table.</p>
<p><em>The <code>mergeAssays</code> is built on top of R’s merge
function (actually, merge from <code>data.table</code>), and therefore
accepts the same arguments to control how the merge is done.</em></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># -- Attach the Hill parameters from the monotherapy profiles to the combination data</span></span>
<span><span class="co"># Extract the monotherapy fits</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">assay</span><span class="op">(</span><span class="st">"mono_profiles"</span>, summarize<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>    <span class="va">Rsq</span> <span class="op">&gt;</span> <span class="fl">0.5</span>,  <span class="co"># filter to only good quality fits</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span>, <span class="st">"HS"</span>, <span class="st">"E_inf"</span>, <span class="st">"EC50"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span></span>
<span>  <span class="va">NCI_TRE</span><span class="op">$</span><span class="va">mono_fits</span></span>
<span></span>
<span><span class="co"># Extract the combination response observations</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">assay</span><span class="op">(</span><span class="st">"response"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">treatment2id</span> <span class="op">!=</span> <span class="st">""</span><span class="op">)</span> <span class="op">-&gt;</span></span>
<span>  <span class="va">NCI_TRE</span><span class="op">$</span><span class="va">combo_response</span></span>
<span></span>
<span><span class="co"># Merge with the fits from the monotherapy experiments</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mergeAssays</span><span class="op">(</span></span>
<span>    x<span class="op">=</span><span class="st">"combo_response"</span>,</span>
<span>    y<span class="op">=</span><span class="st">"mono_fits"</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    sort <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># workaround to not have weird DT keys as a result.</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mergeAssays</span><span class="op">(</span></span>
<span>    x<span class="op">=</span><span class="st">"combo_response"</span>,</span>
<span>    y<span class="op">=</span><span class="st">"mono_fits"</span>,</span>
<span>    by.x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment2id"</span>, <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    by.y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    suffixes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"_1"</span>, <span class="st">"_2"</span><span class="op">)</span>,</span>
<span>    sort <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># workaround to not have weird DT keys as a result.</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span></span>
<span>  <span class="va">NCI_TRE</span></span>
<span><span class="op">(</span><span class="va">NCI_TRE</span><span class="op">$</span><span class="va">combo_response</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                 treatment1id treatment2id treatment1dose</span></span>
<span><span class="co">##      1: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">##      2: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">##      3: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">##      4: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">##      5: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">##     ---                                                                 </span></span>
<span><span class="co">## 127310:                           Vorinostat  Vincristine            1.0</span></span>
<span><span class="co">## 127311:                           Vorinostat  Vincristine            1.0</span></span>
<span><span class="co">## 127312:                           Vorinostat  Vincristine            1.0</span></span>
<span><span class="co">## 127313:                           Vorinostat  Vincristine            1.0</span></span>
<span><span class="co">## 127314:                           Vorinostat  Vincristine            1.0</span></span>
<span><span class="co">##         treatment2dose tech_rep sampleid bio_rep CONC1 CONC2 CONCINDEX2   NSC2</span></span>
<span><span class="co">##      1:           0.03        1    786-O       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">##      2:           0.03        1    A-498       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">##      3:           0.03        1    A-549       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">##      4:           0.03        1     ACHN       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">##      5:           0.03        1 CCRF-CEM       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">##     ---                                                                       </span></span>
<span><span class="co">## 127310:           0.10        1       SR       1 1e-06 1e-07          3  67574</span></span>
<span><span class="co">## 127311:           0.10        1    SW620       1 1e-06 1e-07          3  67574</span></span>
<span><span class="co">## 127312:           0.10        1    T-47D       1 1e-06 1e-07          3  67574</span></span>
<span><span class="co">## 127313:           0.10        1 UACC-257       1 1e-06 1e-07          3  67574</span></span>
<span><span class="co">## 127314:           0.10        1  UACC-62       1 1e-06 1e-07          3  67574</span></span>
<span><span class="co">##          CELLNAME CELLNBR                      PANEL PANELNBR  viability</span></span>
<span><span class="co">##      1:     786-0      18               Renal Cancer        9 101.481127</span></span>
<span><span class="co">##      2:      A498      13               Renal Cancer        9 103.771525</span></span>
<span><span class="co">##      3: A549/ATCC       4 Non-Small Cell Lung Cancer        1  93.210581</span></span>
<span><span class="co">##      4:      ACHN      23               Renal Cancer        9 100.015651</span></span>
<span><span class="co">##      5:  CCRF-CEM       3                   Leukemia        7 106.627710</span></span>
<span><span class="co">##     ---                                                                 </span></span>
<span><span class="co">## 127310:        SR      19                   Leukemia        7   2.297507</span></span>
<span><span class="co">## 127311:    SW-620       9               Colon Cancer        4  39.885949</span></span>
<span><span class="co">## 127312:     T-47D      14              Breast Cancer        5  40.278674</span></span>
<span><span class="co">## 127313:  UACC-257      21                   Melanoma       10  47.402650</span></span>
<span><span class="co">## 127314:   UACC-62      20                   Melanoma       10  36.644778</span></span>
<span><span class="co">##         PERCENTGROWTH PERCENTGROWTHNOTZ EXPECTEDGROWTH TESTVALUE CONTROLVALUE</span></span>
<span><span class="co">##      1:        101.75            101.48        100.000   6951240      6849786</span></span>
<span><span class="co">##      2:        104.84            103.77        100.000   4182200      4030200</span></span>
<span><span class="co">##      3:         92.34             93.21         95.400   5888640      6317566</span></span>
<span><span class="co">##      4:        100.02            100.02         98.098   6607800      6606766</span></span>
<span><span class="co">##      5:        107.76            106.63        100.000  13606520     12760773</span></span>
<span><span class="co">##     ---                                                                      </span></span>
<span><span class="co">## 127310:        -77.71              2.30        -82.190    199440      8680713</span></span>
<span><span class="co">## 127311:         32.22             39.89         14.564   7887720     19775686</span></span>
<span><span class="co">## 127312:          4.38             40.28          4.133   1612280      4002813</span></span>
<span><span class="co">## 127313:          9.23             47.40          5.886   6928760     14616820</span></span>
<span><span class="co">## 127314:        -25.42             36.64        -15.760   4056120     11068753</span></span>
<span><span class="co">##         TZVALUE      HS_1    E_inf_1     EC50_1 HS_2   E_inf_2      EC50_2</span></span>
<span><span class="co">##      1: 1042043 1.0000000 0.08977556  50.000000    1 0.2720558  3.00000000</span></span>
<span><span class="co">##      2:  889724 1.0000000 0.56096579 500.000000    1 0.4647826 30.00000000</span></span>
<span><span class="co">##      3:  716463 1.0000000 0.15732265  50.000000    1 0.2215019  3.00000000</span></span>
<span><span class="co">##      4: 1158169 1.0000000 0.01729107  50.000000    1 0.2172949  3.00000000</span></span>
<span><span class="co">##      5: 1863442 1.0000000 0.12852970  50.000000    1 0.2407407  3.00000000</span></span>
<span><span class="co">##     ---                                                                   </span></span>
<span><span class="co">## 127310:  894742 1.0000000 0.17711243   2.000000    4 0.1928192  0.01523461</span></span>
<span><span class="co">## 127311: 2235554 1.0000000 0.07429719   1.000000    1 0.4680214  0.03000000</span></span>
<span><span class="co">## 127312: 1502784 4.0000000 0.36282782   1.429975    1 0.4751067  0.03000000</span></span>
<span><span class="co">## 127313: 6146751 0.9672031 0.26448079   1.146777    1 0.5252780  0.03000000</span></span>
<span><span class="co">## 127314: 5438488 1.0000000 0.17486339   2.000000    1 0.6472657  0.03000000</span></span></code></pre>
<p>Finally, we can make predictions for each of the measured combination
doses using the corresponding fit parameters and the Hill Curve models.
This step is necessary when interpolating between monotherapy doses, but
is often done in any case, as the predicted viabilities can be less
noisy than each individual monotherapy measurement.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># predict viability at the combo doses from the Hill curves fit to monotherapy data</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">endoaggregate</span><span class="op">(</span></span>
<span>    <span class="op">{</span></span>
<span>      <span class="va">viability_1</span> <span class="op">&lt;-</span> <span class="fu">PharmacoGx</span><span class="fu">:::</span><span class="fu">.Hill</span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">treatment1dose</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">HS_1</span>, <span class="va">E_inf_1</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">EC50_1</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>      <span class="va">viability_2</span> <span class="op">&lt;-</span> <span class="fu">PharmacoGx</span><span class="fu">:::</span><span class="fu">.Hill</span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">treatment2dose</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">HS_2</span>, <span class="va">E_inf_2</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">EC50_2</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        viability_1 <span class="op">=</span> <span class="va">viability_1</span>,</span>
<span>        viability_2 <span class="op">=</span> <span class="va">viability_2</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">}</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"combo_response"</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"treatment1dose"</span>, <span class="st">"treatment2dose"</span>,</span>
<span>    <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    enlist <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span> <span class="va">NCI_TRE</span></span></code></pre></div>
<p>Last, we also put the measured combination viabilily onto a
fractional, not percentage scale.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">NCI_TRE</span><span class="op">$</span><span class="va">combo_response</span><span class="op">$</span><span class="va">viability</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">NCI_TRE</span><span class="op">$</span><span class="va">combo_response</span><span class="op">$</span><span class="va">viability</span> <span class="op">/</span> <span class="fl">100</span></span></code></pre></div>
<p>Finally, we can talk about the different models to compute the
predicted response from monotherapy, and thereby compute estimates of
synergy or antagonism.</p>
</div>
<div class="section level3">
<h3 id="highest-single-agent-highest-single-effect">Highest Single Agent / Highest Single Effect<a class="anchor" aria-label="anchor" href="#highest-single-agent-highest-single-effect"></a>
</h3>
<p>The simplest model that one can assume for the response of drugs in
combination is the Highest Single Agent model. This model predicts that
the result of treating with two drugs is the same as the highest
response (or lowest viability) seen by treating with either drug
independently. Mathematically, this is equivalent to:</p>
<p><span class="math display">\[HSA((r_1,x_1), (r_2,x_2)) =
\mbox{max}(r_1, r_2)\]</span></p>
<p>For viability values, this flips to a minimum.</p>
<p>Intuitively, this model can be interpreted as assuming that one
compound is at the maximum possible effect, and the other is treated as
a potentiator, contributing little to the response itself, but allowing
the maximum response of the first drug to be increased. Practically,
this case is not very often observed during <em>in vitro</em>
dose-response experiments, and the highest single agent model is often a
very weak baseline from which to compute synergy.</p>
<div class="section level4">
<h4 id="computing-hsa-reference-viabilities">Computing HSA Reference Viabilities<a class="anchor" aria-label="anchor" href="#computing-hsa-reference-viabilities"></a>
</h4>
<p>As one can imagine, this is very simple to do using the
<code>min</code> or <code>max</code> functions in R.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># -- compute our drug synergy metrics</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">endoaggregate</span><span class="op">(</span></span>
<span>    HSA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">viability_1</span>, <span class="va">viability_2</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"treatment1dose"</span>, <span class="st">"treatment2dose"</span>,</span>
<span>      <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"combo_response"</span>,</span>
<span>    target <span class="op">=</span> <span class="st">"combo_response"</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span> <span class="va">NCI_TRE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">NCI_TRE</span><span class="op">$</span><span class="va">combo_response</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            treatment1id treatment2id treatment1dose</span></span>
<span><span class="co">## 1: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 2: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 3: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 4: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 5: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 6: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">##    treatment2dose tech_rep sampleid bio_rep CONC1 CONC2 CONCINDEX2   NSC2</span></span>
<span><span class="co">## 1:           0.03        1    786-O       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 2:           0.03        1    A-498       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 3:           0.03        1    A-549       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 4:           0.03        1     ACHN       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 5:           0.03        1 CCRF-CEM       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 6:           0.03        1 COLO 205       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">##     CELLNAME CELLNBR                      PANEL PANELNBR viability</span></span>
<span><span class="co">## 1:     786-0      18               Renal Cancer        9 1.0148113</span></span>
<span><span class="co">## 2:      A498      13               Renal Cancer        9 1.0377152</span></span>
<span><span class="co">## 3: A549/ATCC       4 Non-Small Cell Lung Cancer        1 0.9321058</span></span>
<span><span class="co">## 4:      ACHN      23               Renal Cancer        9 1.0001565</span></span>
<span><span class="co">## 5:  CCRF-CEM       3                   Leukemia        7 1.0662771</span></span>
<span><span class="co">## 6:  COLO 205      10               Colon Cancer        4 0.9205163</span></span>
<span><span class="co">##    PERCENTGROWTH PERCENTGROWTHNOTZ EXPECTEDGROWTH TESTVALUE CONTROLVALUE</span></span>
<span><span class="co">## 1:        101.75            101.48        100.000   6951240      6849786</span></span>
<span><span class="co">## 2:        104.84            103.77        100.000   4182200      4030200</span></span>
<span><span class="co">## 3:         92.34             93.21         95.400   5888640      6317566</span></span>
<span><span class="co">## 4:        100.02            100.02         98.098   6607800      6606766</span></span>
<span><span class="co">## 5:        107.76            106.63        100.000  13606520     12760773</span></span>
<span><span class="co">## 6:         90.74             92.05         86.870   5822480      6325233</span></span>
<span><span class="co">##    TZVALUE HS_1    E_inf_1 EC50_1 HS_2    E_inf_2 EC50_2 viability_1</span></span>
<span><span class="co">## 1: 1042043    1 0.08977556     50    1 0.27205581      3   0.9963736</span></span>
<span><span class="co">## 2:  889724    1 0.56096579    500    1 0.46478261     30   0.9998245</span></span>
<span><span class="co">## 3:  716463    1 0.15732265     50    1 0.22150189      3   0.9966427</span></span>
<span><span class="co">## 4: 1158169    1 0.01729107     50    1 0.21729490      3   0.9960848</span></span>
<span><span class="co">## 5: 1863442    1 0.12852970     50    1 0.24074074      3   0.9965280</span></span>
<span><span class="co">## 6:  895972    1 0.07365439     50    1 0.02186312      3   0.9963094</span></span>
<span><span class="co">##    viability_2       HSA</span></span>
<span><span class="co">## 1:   0.9927926 0.9927926</span></span>
<span><span class="co">## 2:   0.9994653 0.9994653</span></span>
<span><span class="co">## 3:   0.9922921 0.9922921</span></span>
<span><span class="co">## 4:   0.9922504 0.9922504</span></span>
<span><span class="co">## 5:   0.9924826 0.9924826</span></span>
<span><span class="co">## 6:   0.9903155 0.9903155</span></span></code></pre>
<p>For convenience,
<code>PharmacoGx::computeHSA(viability_1, viability_2)</code> is a more
readable alias for computing the HSA reference viability.</p>
</div>
</div>
<div class="section level3">
<h3 id="bliss-independence">Bliss Independence<a class="anchor" aria-label="anchor" href="#bliss-independence"></a>
</h3>
<p>Bliss Independence can be understood by making the assumption that
cell death from treatment with a drugs is a stochastic process with the
probability of cell death being proportional to the response observed in
the experiment. The model than assumes that the cell death events from
the two drugs are independent, for example because they occur through
completely separate pathways. As such, the model predicts that adding a
second drug will kill the same fraction of cells as it kills on its own,
and therefore the fractions of cells remaining alive are caculated by
multplying the fraction of cells alive after treatment with drug 1 in
monotherapy, with the fraction of cells alive after treatment with drug
2 in monotherapy.</p>
<div class="figure">
<img src="BlissDiagram.png" alt=""><p class="caption">Diagram of Bliss Independence Assumption. Bliss
independence assumes a constant fraction of cells is killed
independently by each drug. In this diagram, the orange drug leaves 40%
of the cells alive, and the blue drug leaves 60% of the cells alive.
Therefore, Bliss predicts that the two drugs in combination will leave
24% of the cells alive.</p>
</div>
<p>Mathematically, that works out to:</p>
<p><span class="math display">\[BLISS((v_1,x_1), (v_2,x_2)) = v_1
v_2\]</span></p>
<p>In response units this is:</p>
<p><span class="math display">\[BLISS((r_1,x_1), (r_2,x_2)) = r_1 + r_2
- r_1*r_2\]</span></p>
<p>This model is always going to be more strict that HSA (as long as
response is between 0 and 1), and is usually a more practical assumption
to make when searching for drug synergy. Intuitively, one can think of
it as modeling the case where the two drugs have separate mechanisms,
with synergy signifiying a biological interaction that makes them more
effective than expected.</p>
<div class="section level4">
<h4 id="computing-bliss-reference-viabilities">Computing Bliss Reference Viabilities<a class="anchor" aria-label="anchor" href="#computing-bliss-reference-viabilities"></a>
</h4>
<p>This is also quite simple to do, as we just use the product of the
monotherapy viabilities.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># -- compute our drug synergy metrics</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">endoaggregate</span><span class="op">(</span></span>
<span>    Bliss <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/prod.html" class="external-link">prod</a></span><span class="op">(</span><span class="va">viability_1</span>, <span class="va">viability_2</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"treatment1dose"</span>, <span class="st">"treatment2dose"</span>,</span>
<span>      <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    assay<span class="op">=</span><span class="st">"combo_response"</span>,</span>
<span>    target <span class="op">=</span> <span class="st">"combo_response"</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span> <span class="va">NCI_TRE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">NCI_TRE</span><span class="op">$</span><span class="va">combo_response</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            treatment1id treatment2id treatment1dose</span></span>
<span><span class="co">## 1: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 2: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 3: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 4: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 5: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 6: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">##    treatment2dose tech_rep sampleid bio_rep CONC1 CONC2 CONCINDEX2   NSC2</span></span>
<span><span class="co">## 1:           0.03        1    786-O       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 2:           0.03        1    A-498       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 3:           0.03        1    A-549       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 4:           0.03        1     ACHN       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 5:           0.03        1 CCRF-CEM       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 6:           0.03        1 COLO 205       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">##     CELLNAME CELLNBR                      PANEL PANELNBR viability</span></span>
<span><span class="co">## 1:     786-0      18               Renal Cancer        9 1.0148113</span></span>
<span><span class="co">## 2:      A498      13               Renal Cancer        9 1.0377152</span></span>
<span><span class="co">## 3: A549/ATCC       4 Non-Small Cell Lung Cancer        1 0.9321058</span></span>
<span><span class="co">## 4:      ACHN      23               Renal Cancer        9 1.0001565</span></span>
<span><span class="co">## 5:  CCRF-CEM       3                   Leukemia        7 1.0662771</span></span>
<span><span class="co">## 6:  COLO 205      10               Colon Cancer        4 0.9205163</span></span>
<span><span class="co">##    PERCENTGROWTH PERCENTGROWTHNOTZ EXPECTEDGROWTH TESTVALUE CONTROLVALUE</span></span>
<span><span class="co">## 1:        101.75            101.48        100.000   6951240      6849786</span></span>
<span><span class="co">## 2:        104.84            103.77        100.000   4182200      4030200</span></span>
<span><span class="co">## 3:         92.34             93.21         95.400   5888640      6317566</span></span>
<span><span class="co">## 4:        100.02            100.02         98.098   6607800      6606766</span></span>
<span><span class="co">## 5:        107.76            106.63        100.000  13606520     12760773</span></span>
<span><span class="co">## 6:         90.74             92.05         86.870   5822480      6325233</span></span>
<span><span class="co">##    TZVALUE HS_1    E_inf_1 EC50_1 HS_2    E_inf_2 EC50_2 viability_1</span></span>
<span><span class="co">## 1: 1042043    1 0.08977556     50    1 0.27205581      3   0.9963736</span></span>
<span><span class="co">## 2:  889724    1 0.56096579    500    1 0.46478261     30   0.9998245</span></span>
<span><span class="co">## 3:  716463    1 0.15732265     50    1 0.22150189      3   0.9966427</span></span>
<span><span class="co">## 4: 1158169    1 0.01729107     50    1 0.21729490      3   0.9960848</span></span>
<span><span class="co">## 5: 1863442    1 0.12852970     50    1 0.24074074      3   0.9965280</span></span>
<span><span class="co">## 6:  895972    1 0.07365439     50    1 0.02186312      3   0.9963094</span></span>
<span><span class="co">##    viability_2       HSA     Bliss</span></span>
<span><span class="co">## 1:   0.9927926 0.9927926 0.9891924</span></span>
<span><span class="co">## 2:   0.9994653 0.9994653 0.9992899</span></span>
<span><span class="co">## 3:   0.9922921 0.9922921 0.9889607</span></span>
<span><span class="co">## 4:   0.9922504 0.9922504 0.9883656</span></span>
<span><span class="co">## 5:   0.9924826 0.9924826 0.9890367</span></span>
<span><span class="co">## 6:   0.9903155 0.9903155 0.9866606</span></span></code></pre>
<p>For convenience,
<code>PharmacoGx::computeBliss(viability_1, viability_2)</code> is a
more readable alias for computing the Bliss reference viability.</p>
</div>
</div>
<div class="section level3">
<h3 id="loewe-additivity">Loewe Additivity<a class="anchor" aria-label="anchor" href="#loewe-additivity"></a>
</h3>
<p>Loewe Additivity comes from a different assumption on what the effect
of adding two drugs should be. While Bliss Independence assumes that the
drugs act independently, for example through different pathways, Loewe
assumes that in the null case, adding <span class="math inline">\(x_2\)</span> concentration of a second drug should
lower the viability as if we added a proportional amount of the same
drug, with that proportion caculated according to the relative efficacy
of the two drugs at the final combination viability. This proportinal
relationship is defined implicity at each predicted value of $v_{loewe}
by the following formula:</p>
<p><span class="math display">\[ \frac{x_1}{X^{loewe}_1} +
\frac{x_2}{X^{loewe}_2} = 1\]</span></p>
<p>where $ X^{loewe}_1 $ is the concentration of drug 1 required to
achieve <span class="math inline">\(v_{loewe}\)</span> on its own, and
similarly for $ X^{loewe}_2$ and drug 2. Translating this to english, it
assumes that we can lower the concentration of drug 1 needed to see the
same effect by replacing the missing dose with a similar amount of drug
2, adjusted for the difference in potency between the two drugs.</p>
<p>Practically, this formula is not very useful for estimating <span class="math inline">\(v_{loewe}\)</span> from actual dose-response data.
However, if we assume that drug response follows a Hill Curve model, we
can define <span class="math inline">\(y_{loewe}\)</span> implicitly
using the parameters <span class="math inline">\(HS_{1}\)</span>, <span class="math inline">\(EC50_{1}\)</span>, and <span class="math inline">\(E_{\infty,1}\)</span>, and <span class="math inline">\(HS_{2}\)</span>, <span class="math inline">\(EC50_{2}\)</span>, and <span class="math inline">\(E_{\infty,2}\)</span> by rearranging the Hill
Curve relationship as follows:</p>
<p><span class="math display">\[ x = EC50 *
(\frac{v-1}{v-E_{\infty}})^{\frac{1}{HS}}\]</span></p>
<p>We can then plug this relationship in to get an emplicit formula for
<span class="math inline">\(v_{loewe}\)</span>:</p>
<p><span class="math display">\[ \frac{x_1}{EC50_1 *
(\frac{v_{loewe}-1}{v_{loewe}-E_{\infty,1}})^{\frac{1}{HS_1}}} +
\frac{x_2}{EC50_2 *
(\frac{v_{loewe}-1}{v_{loewe}-E_{\infty,2}})^{\frac{1}{HS_2}}} =
1\]</span></p>
<p>This formula can be solved numerically to find the value of <span class="math inline">\(v_{loewe}\)</span> for any two particular fitted
curves.</p>
<p>While the Loewe Additivity model can be used to calculate the
expected response at any pair of doses, more commonly, the Loewe
Additivity is assessed in terms of the Combination Index. The
combination index measures the deviation from the Loewe assumption as a
ratio at each point, by computing the Loewe model with the observed
combination viability instead of the reference viability. As it doesn’t
require calculating <span class="math inline">\(v_{loewe}\)</span>, it
is also easier to compute. The formula for the Combination Index given a
Loewe assumption is:</p>
<p><span class="math display">\[ CI_{loewe} =
\frac{x_1}{EC50^1*(\frac{v_c}{E_{\infty}^1-v_c})^{1/HS^1}} +
\frac{x_2}{EC50^2*(\frac{v_c}{E_{\infty}^2-v_c})^{1/HS^2}} \]</span></p>
<div class="figure">
<img src="LoeweDiagram.png" alt=""><p class="caption">Diagram of Loewe Additivity Assumption. Loewe
additivity assumes that when plotting viability as a function of both
concentrations (bottom plots), lines of equal viability (isoboles) will
be straight. Using this assumption, either the expected viability for
each pair of concentrations can be calculated (left), or the combination
index can be computed to assess synergy or antagonism directly
(right).</p>
</div>
<div class="section level4">
<h4 id="computing-loewe-reference-viabilities">Computing Loewe Reference Viabilities<a class="anchor" aria-label="anchor" href="#computing-loewe-reference-viabilities"></a>
</h4>
<p>As mentioned above, the computation of the Loewe reference viability
requires solving a non-linear implicitly defined equation. PharmacoGx
provides the <code>computeLoewe</code> function to do this
calculation.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># -- compute our drug synergy metrics</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">endoaggregate</span><span class="op">(</span></span>
<span>    Loewe <span class="op">=</span> <span class="fu">PharmacoGx</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PharmacoGx/man/computeLoewe.html" class="external-link">computeLoewe</a></span><span class="op">(</span></span>
<span>      <span class="va">treatment1dose</span>, <span class="va">HS_1</span>, <span class="va">E_inf_1</span>, <span class="va">EC50_1</span>,</span>
<span>      <span class="va">treatment2dose</span>, <span class="va">HS_2</span>, <span class="va">E_inf_2</span>, <span class="va">EC50_2</span></span>
<span>    <span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"treatment1dose"</span>, <span class="st">"treatment2dose"</span>,</span>
<span>      <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"combo_response"</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span> <span class="va">NCI_TRE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">NCI_TRE</span><span class="op">$</span><span class="va">combo_response</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            treatment1id treatment2id treatment1dose</span></span>
<span><span class="co">## 1: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 2: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 3: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 4: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 5: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 6: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">##    treatment2dose tech_rep sampleid bio_rep CONC1 CONC2 CONCINDEX2   NSC2</span></span>
<span><span class="co">## 1:           0.03        1    786-O       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 2:           0.03        1    A-498       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 3:           0.03        1    A-549       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 4:           0.03        1     ACHN       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 5:           0.03        1 CCRF-CEM       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">## 6:           0.03        1 COLO 205       1 2e-07 3e-08          1 102816</span></span>
<span><span class="co">##     CELLNAME CELLNBR                      PANEL PANELNBR viability</span></span>
<span><span class="co">## 1:     786-0      18               Renal Cancer        9 1.0148113</span></span>
<span><span class="co">## 2:      A498      13               Renal Cancer        9 1.0377152</span></span>
<span><span class="co">## 3: A549/ATCC       4 Non-Small Cell Lung Cancer        1 0.9321058</span></span>
<span><span class="co">## 4:      ACHN      23               Renal Cancer        9 1.0001565</span></span>
<span><span class="co">## 5:  CCRF-CEM       3                   Leukemia        7 1.0662771</span></span>
<span><span class="co">## 6:  COLO 205      10               Colon Cancer        4 0.9205163</span></span>
<span><span class="co">##    PERCENTGROWTH PERCENTGROWTHNOTZ EXPECTEDGROWTH TESTVALUE CONTROLVALUE</span></span>
<span><span class="co">## 1:        101.75            101.48        100.000   6951240      6849786</span></span>
<span><span class="co">## 2:        104.84            103.77        100.000   4182200      4030200</span></span>
<span><span class="co">## 3:         92.34             93.21         95.400   5888640      6317566</span></span>
<span><span class="co">## 4:        100.02            100.02         98.098   6607800      6606766</span></span>
<span><span class="co">## 5:        107.76            106.63        100.000  13606520     12760773</span></span>
<span><span class="co">## 6:         90.74             92.05         86.870   5822480      6325233</span></span>
<span><span class="co">##    TZVALUE HS_1    E_inf_1 EC50_1 HS_2    E_inf_2 EC50_2 viability_1</span></span>
<span><span class="co">## 1: 1042043    1 0.08977556     50    1 0.27205581      3   0.9963736</span></span>
<span><span class="co">## 2:  889724    1 0.56096579    500    1 0.46478261     30   0.9998245</span></span>
<span><span class="co">## 3:  716463    1 0.15732265     50    1 0.22150189      3   0.9966427</span></span>
<span><span class="co">## 4: 1158169    1 0.01729107     50    1 0.21729490      3   0.9960848</span></span>
<span><span class="co">## 5: 1863442    1 0.12852970     50    1 0.24074074      3   0.9965280</span></span>
<span><span class="co">## 6:  895972    1 0.07365439     50    1 0.02186312      3   0.9963094</span></span>
<span><span class="co">##    viability_2       HSA     Bliss     Loewe</span></span>
<span><span class="co">## 1:   0.9927926 0.9927926 0.9891924 0.9892436</span></span>
<span><span class="co">## 2:   0.9994653 0.9994653 0.9992899 0.9993076</span></span>
<span><span class="co">## 3:   0.9922921 0.9922921 0.9889607 0.9889720</span></span>
<span><span class="co">## 4:   0.9922504 0.9922504 0.9883656 0.9884240</span></span>
<span><span class="co">## 5:   0.9924826 0.9924826 0.9890367 0.9890752</span></span>
<span><span class="co">## 6:   0.9903155 0.9903155 0.9866606 0.9867043</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="bonus-zip">BONUS: ZIP<a class="anchor" aria-label="anchor" href="#bonus-zip"></a>
</h3>
<p>The Zero interaction Potency model was recently introduced with the
goal of capturing the ideas behind both the Bliss and Loewe assumptions.
As the name implies the model assumes that the two drugs do not
“potentiate” each other, meaning, that the only difference between the
dose response curves of Drug 1 with and without Drug 2 present is that
the presence of Drug 2 changes the baseline response, without affecting
the potency of Drug 1. While the ZIP model can be extended to 3 and 4
parameter Hill Curve models, it is simplest to understand and explain
using the 2 parameter Hill Curve, which is equivalent to the case where
<span class="math inline">\(E_{\infty}=0\)</span>.</p>
<p>The authors were then able to show that the ZIP model can be derived
from multiplying the two fitted Dose-Response curves together.</p>
<p><span class="math display">\[ZIP((v_1,x_1), (v_2,x_2)) =
(E_{\infty}^1 +
\frac{1-E_{infty}^1}{1+(\frac{x_1}{EC50^1})^{HS^1}})*(E_{\infty}^2 +
\frac{1-E_{infty}^2}{1+(\frac{x_2}{EC50^2})^{HS^2}})\]</span></p>
<p>While the formula for computing the ZIP reference value at each point
works out to being the Bliss formula applied to the fitted curves, the
ZIP assumptions suggest a new method to assess syngergy or antagosism at
each point. Building on the assumption that combination without synergy
should not change the potency of each particular drug, the ZIP <span class="math inline">\(\delta\)</span> score defines a measure comparing
the difference between the fitted curve for each drug with and without
the presence of the second drug, taking the average of differences for
the two drugs.</p>
<p>Mathematically, ZIP defines two values: <span class="math display">\[
v_c^{1 \leftarrow 2}((v_1,x_1), (v_2,x_2)) =
\frac{v_2}{1+(\frac{x_1}{EC50^{1 \leftarrow 2}})^{HS^{1 \leftarrow
2}}}\]</span> <span class="math display">\[ v_c^{2 \leftarrow
1}((v_1,x_1), (v_2,x_2)) = \frac{v_1}{1+(\frac{x_1}{EC50^{2 \leftarrow
1}})^{HS^{2 \leftarrow 1}}}\]</span></p>
<p>In this notation, {<span class="math inline">\(1 \leftarrow
2\)</span>} means that we fit the Hill Curve for drug 1 after the
addition of drug 2 at the dose <span class="math inline">\(x_2\)</span>.
Intuitively, we adjust for the observed effect of adding <span class="math inline">\(x_2\)</span> of drug 2 (<span class="math inline">\(v_2\)</span>), and then treat drug 1 as if it is a
monotherapy experiment. The <span class="math inline">\(\delta\)</span>
score is then defined as the difference of the viability predicted by
this curve and the ZIP predicted viability, averaged over treating each
of the drugs as the “main drug”:</p>
<p><span class="math display">\[\delta((v_1,x_1), (v_2,x_2)) =
\frac{v_c^{1 \leftarrow 2} - v_{ZIP}}{2} + \frac{v_c^{1 \leftarrow 2} -
v_{ZIP}}{2}\]</span></p>
<p>For more information on the ZIP model, we refer the reader to the
original publication <span class="citation">(Yadav et al. 2015)</span>.
Alternatively, the <em><a href="https://bioconductor.org/packages/3.16/synergyfinder" class="external-link">synergyfinder</a></em>
package contains vignettes and the reference implementation for ZIP
calculations.</p>
</div>
<div class="section level3">
<h3 id="picking-a-null-drug-combination-model">Picking a Null Drug Combination Model<a class="anchor" aria-label="anchor" href="#picking-a-null-drug-combination-model"></a>
</h3>
<p>Given the fact that there is no universal consensus on which null
model is correct in all cases, when analyzing drug combination data, a
choice must be made by the analyst. Ideally, this choice would be made
according to reasonable assumptions about the drugs and their mechanisms
of action. For example, for two compounds which inhibit targets
converging on the same pathway, the Loewe model may be most appropriate,
whereas for drug combinations with completely unrelated mechanisms,
Bliss may be a more natural choice.</p>
<p>However, practical considerations may also limit your choices. For
both the Loewe and ZIP approaches, one needs enough points to be
measured to be able to fit Hill Curves to the available data. For Loewe,
this needs to be done only in monotherapy, while for ZIP, both
monotherapy and combination data must have sufficient points to fit a
curve. On the other hand, both HSA and Bliss can be computed from the
raw measurements of monotherapy viability, not requiring any models of
the dose-response data to be fit.</p>
<p>In general, we would recommend using a model which makes stronger
assumptions than HSA, which is the weakest of the presented models. If
the number of tested doses is low (&lt;6), the Bliss model is probably
the reference of choice, even though it is known to be quite strict.
When possible to calculate, the ZIP approach has advantages in a
high-throughput setting, where it would be difficult to pick between
Bliss and Loewe a-priori for each individual combination.</p>
</div>
<div class="section level3">
<h3 id="quantifying-synergy-or-antagonism-compared-to-a-null-model">Quantifying Synergy or Antagonism Compared to a Null Model<a class="anchor" aria-label="anchor" href="#quantifying-synergy-or-antagonism-compared-to-a-null-model"></a>
</h3>
<p>Once a null model is chosen, the expected combination viability or
response is compared to the observed value, as mentioned above. For all
methods, this can be done using either a difference, or a ratio. In
addition, the Loewe model directly facilitates calculating the
Combination Index described above, and the ZIP model defines the <span class="math inline">\(\delta\)</span> score for quantifying synergy. The
main practical consideration when choosing how to compare the
observations to the null is the scale on which results will be
interpreted or further analyzed. Taking the ratio of <span class="math inline">\(v_c/v_{model}\)</span>, or using the Combination
Index for Loewe leads to values that lie in <span class="math inline">\((0, \infty]\)</span> (with 1 as the reference for
no syngergy or antagonism), similar to fold changes. On the other hand,
both differences and the ZIP <span class="math inline">\(\delta\)</span>-score lie on a bounded scale <span class="math inline">\([-1,1]\)</span>, with 0 as the reference.</p>
<p>If the goal is simply to identify whether there was synergy or
antagonism observed at a certain dose, the last step would be to
visualize the results, and examine the extent of deviation from the null
model. For this, the scale chosen is a matter of preference. However, if
the downstream goal is biomarker discovery using a summary of synergy
for each model tested with a particular combination, if the Combination
Index or a ratio is used, it may be more appropriate to analyze the
results on a log-scale, as otherwise the range is asymmetric around the
reference.</p>
<p>In practice, our personal preference is to use the absolute deviation
from the Bliss predicted value for downstream analysis.</p>
<p><em>Note that we also skip the ZIP delta score computation in the
interest of time.</em></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># -- compute combination index and score</span></span>
<span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">endoaggregate</span><span class="op">(</span></span>
<span>    HSA_CI <span class="op">=</span> <span class="va">viability</span> <span class="op">/</span> <span class="va">HSA</span>,</span>
<span>    HSA_score <span class="op">=</span> <span class="va">HSA</span> <span class="op">-</span> <span class="va">viability</span>,</span>
<span>    Bliss_CI <span class="op">=</span> <span class="va">viability</span> <span class="op">/</span> <span class="va">Bliss</span>,</span>
<span>    Bliss_score <span class="op">=</span> <span class="va">Bliss</span> <span class="op">-</span> <span class="va">viability</span>,</span>
<span>    Loewe_CI <span class="op">=</span> <span class="op">(</span></span>
<span>      <span class="op">(</span><span class="va">treatment1dose</span> <span class="op">/</span></span>
<span>        <span class="va">EC50_1</span> <span class="op">*</span> <span class="op">(</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">viability_1</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">viability_1</span> <span class="op">-</span> <span class="va">E_inf_1</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="va">HS_1</span><span class="op">)</span></span>
<span>      <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="op">(</span><span class="va">treatment2dose</span> <span class="op">/</span></span>
<span>          <span class="va">EC50_2</span> <span class="op">*</span> <span class="op">(</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">viability_2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">viability_2</span> <span class="op">-</span> <span class="va">E_inf_2</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="va">HS_2</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>        <span class="op">)</span>,</span>
<span>    Loewe_score <span class="op">=</span> <span class="va">Loewe</span> <span class="op">-</span> <span class="va">viability</span>,</span>
<span>    viability <span class="op">=</span> <span class="va">viability</span>,</span>
<span>    viability_1 <span class="op">=</span> <span class="va">viability_1</span>,</span>
<span>    viability_2 <span class="op">=</span> <span class="va">viability_2</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"treatment1dose"</span>, <span class="st">"treatment2dose"</span>,</span>
<span>      <span class="st">"sampleid"</span>, <span class="st">"bio_rep"</span><span class="op">)</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"combo_response"</span>,</span>
<span>    target <span class="op">=</span> <span class="st">"combo_scores"</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span></span>
<span>  <span class="va">NCI_TRE</span></span>
<span></span>
<span><span class="va">combo_scores</span> <span class="op">&lt;-</span> <span class="va">NCI_TRE</span><span class="op">$</span><span class="va">combo_scores</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combo_scores</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            treatment1id treatment2id treatment1dose</span></span>
<span><span class="co">## 1: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 2: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 3: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 4: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 5: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">## 6: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine            0.2</span></span>
<span><span class="co">##    treatment2dose sampleid bio_rep    HSA_CI    HSA_score  Bliss_CI Bliss_score</span></span>
<span><span class="co">## 1:           0.03    786-O       1 1.0221785 -0.022018634 1.0258988 -0.02561889</span></span>
<span><span class="co">## 2:           0.03    A-498       1 1.0382704 -0.038249933 1.0384527 -0.03842538</span></span>
<span><span class="co">## 3:           0.03    A-549       1 0.9393462  0.060186284 0.9425105  0.05685488</span></span>
<span><span class="co">## 4:           0.03     ACHN       1 1.0079678 -0.007906062 1.0119297 -0.01179090</span></span>
<span><span class="co">## 5:           0.03 CCRF-CEM       1 1.0743535 -0.073794516 1.0780966 -0.07724041</span></span>
<span><span class="co">## 6:           0.03 COLO 205       1 0.9295182  0.069799189 0.9329614  0.06614431</span></span>
<span><span class="co">##    Loewe_CI Loewe_score viability viability_1 viability_2</span></span>
<span><span class="co">## 1: 1.16e-04 -0.02556769 1.0148113   0.9963736   0.9927926</span></span>
<span><span class="co">## 2: 1.16e-06 -0.03840768 1.0377152   0.9998245   0.9994653</span></span>
<span><span class="co">## 3: 1.16e-04  0.05686616 0.9321058   0.9966427   0.9922921</span></span>
<span><span class="co">## 4: 1.16e-04 -0.01173252 1.0001565   0.9960848   0.9922504</span></span>
<span><span class="co">## 5: 1.16e-04 -0.07720193 1.0662771   0.9965280   0.9924826</span></span>
<span><span class="co">## 6: 1.16e-04  0.06618805 0.9205163   0.9963094   0.9903155</span></span></code></pre>
<p>Once we have a measure of synergy calculated for each dose
combination in a particular experiment, we often would like to visualize
the whole surface of drug-dose combinations to understand how synergy
depends on dosage.</p>
<p>Below, we plot the Bliss score for an antagonistic combination of
drugs, Mitotane and Paclitaxel, on the A-549 cell line. If we recall the
formula for the Bliss score above, we took <span class="math inline">\(v_{model} - v_c\)</span>, meaning that in this
case, negative numbers indicate antagonism, as there is more viability
observed than expected.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">combo_scores</span><span class="op">[</span></span>
<span>    <span class="va">treatment1id</span> <span class="op">==</span> <span class="st">"Mitotane"</span> <span class="op">&amp;</span> <span class="va">treatment2id</span> <span class="op">==</span> <span class="st">"Paclitaxel"</span> <span class="op">&amp;</span> <span class="va">sampleid</span> <span class="op">==</span> <span class="st">"A-549"</span></span>
<span>    <span class="op">]</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">treatment1dose</span>, <span class="va">treatment2dose</span>, fill <span class="op">=</span> <span class="va">Bliss_score</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Mitotane"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Paclitaxel"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><img src="PGxWorkshop_files/figure-html/bliss_score_plot-1.png" width="700"></p>
<p>Note that here, it seems that antagonism is exacerbated at higher
doses, suggesting a dose-dependant mechanism for antagonism.</p>
</div>
<div class="section level3">
<h3 id="summarizing-synergy-matrices-into-single-values">Summarizing Synergy Matrices into Single Values<a class="anchor" aria-label="anchor" href="#summarizing-synergy-matrices-into-single-values"></a>
</h3>
<p>Often, we will want to summarize a whole synergy surface into a
single synergy score for a drug combination per cell-line, for example,
if we want to discover biomarkers of synergy. There are several methods
that this can be done, the most common of which is taking the average
value of your score matrix for the combination. If your doses are
equally spaced (on the log scale), this will give a measure of Volume
Under the Surface.</p>
<p>For some experiments, you may want to be a bit more careful, and not
select the average of the whole surface. For example, in the experiment
above, we see very strong antagonism when both doses of the drugs are
high. However, this may not be a very interesting regime, as often one
of the goals of drug combination therapy is to limit the dose of any
single drug used to reduce associated toxicities. Therefore, an option
could be to summarize only over a dose-range of interest.</p>
<p>If you find yourself in a situation where synergy is observed only in
a narrow dose range, the most extreme deviation from the reference
(deviation from 0 or 1, depending on use of additive or fold-change
metrics) may instead be more appropriate.</p>
<p>For our demo, we will summarize the combination profiles we have
calculated into a Mean Bliss Score per experiment.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">NCI_TRE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span></span>
<span>    Mean_Bliss <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Bliss_score</span><span class="op">)</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"combo_scores"</span>,</span>
<span>    by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1id"</span>, <span class="st">"treatment2id"</span>, <span class="st">"sampleid"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">-&gt;</span></span>
<span>  <span class="va">mean_bliss_score</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                treatment1id treatment2id sampleid   Mean_Bliss</span></span>
<span><span class="co">##     1: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine    786-O -0.038395834</span></span>
<span><span class="co">##     2: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine    A-498  0.063834787</span></span>
<span><span class="co">##     3: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine    A-549 -0.048417317</span></span>
<span><span class="co">##     4: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine     ACHN -0.039941211</span></span>
<span><span class="co">##     5: 1,3-Bis(2-chloroethyl)-1-nitrosourea  Azacitidine CCRF-CEM -0.018621532</span></span>
<span><span class="co">##    ---                                                                        </span></span>
<span><span class="co">## 14142:                           Vorinostat  Vincristine       SR  0.170580191</span></span>
<span><span class="co">## 14143:                           Vorinostat  Vincristine    SW620 -0.078155666</span></span>
<span><span class="co">## 14144:                           Vorinostat  Vincristine    T-47D -0.006124021</span></span>
<span><span class="co">## 14145:                           Vorinostat  Vincristine UACC-257 -0.025048742</span></span>
<span><span class="co">## 14146:                           Vorinostat  Vincristine  UACC-62  0.099422703</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="biomarker-discovery">Biomarker Discovery:<a class="anchor" aria-label="anchor" href="#biomarker-discovery"></a>
</h2>
<div class="section level3">
<h3 id="monotherapy-response">Monotherapy response<a class="anchor" aria-label="anchor" href="#monotherapy-response"></a>
</h3>
<p>PharmacoGx makes an initial analysis for univariate biomarker
discovery very simple, by providing the <code>drugSensitivitySig</code>
function which will compute the association between molecular features
and drug response summaries (such as AAC) across the cell lines in a
PharmacoSet.</p>
<p>There are two approaches implemented in PharmacoGx for measuring the
association between molecular features and drug response. The first,
older approach is to use a linear model to estimate the effect of the
molecular feature measurements on the drug response summary, and an
analysis of variance F-test to test the significance of the molecular
feature in predicting drug response over a Null model. If multiple
tissue types are present in the dataset, the Null model includes the
tissue as an explanatory features, otherwise, it is just the intercept.
This approach is most flexible, allowing for both continous and binary
measurements of drug response, as well as continous, categorical and
binary molecular features. However, the estimate of the effect returned
is a standardized regression coefficient, which may be difficult to
interpret and compare between biomarkers.</p>
<p>The second approach is to use a Pearson Correlation, or a Partial
Pearson Correlation corrected for tissue type to test for associations.
Currently, this is our preferred approach for two reasons:</p>
<ol style="list-style-type: decimal">
<li>The returned effect size estimate will be on the scale of a Pearson
Correlation, which is familiar to most scientists and comparable between
molecular features (as long as the sample sizes do not vary too
drastically).</li>
<li>Currently, PharmacoGx has implemented efficient permutation testing
of Pearson Correlations for significance assessment, in addition to the
regular asymptotic formulas for inferring significance.</li>
</ol>
<p>As our focus is on drug combination analysis in this workshop, we
won’t be demonstrating the usage of <code>drugSensitivitySig</code> to
compute monotherapy associations with drug response. Rather, we skip
directly to computing associations with drug synergy. An example of
monotherapy analysis can be found in the <em><a href="https://bioconductor.org/packages/3.16/PharmacoGx" class="external-link">PharmacoGx</a></em>
vignette: <code>vignette("PharmacoGx")</code>.</p>
</div>
<div class="section level3">
<h3 id="combination-synergy">Combination Synergy<a class="anchor" aria-label="anchor" href="#combination-synergy"></a>
</h3>
<p>The <code>drugSensitivitySig</code> function can be used to find
biomarkers for synergy, by passing in a matrix of synergy values for
each drug combination and cell line.</p>
<p>Above, we had already summarized our synergy surfaces into a single
mean value for each experiment. Now, we need to reshape the “long”
format table into a table of drug combinations x cell lines.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create a matrix of drug combo x cell lines.</span></span>
<span><span class="va">synergy_summaries</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/dcast.data.table.html" class="external-link">dcast</a></span><span class="op">(</span></span>
<span>  <span class="va">mean_bliss_score</span>,</span>
<span>  <span class="va">treatment1id</span> <span class="op">+</span> <span class="va">treatment2id</span> <span class="op">~</span> <span class="va">sampleid</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Name our rows something sensible related to the drug combinations.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">synergy_summaries</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>    <span class="va">synergy_summaries</span><span class="op">$</span><span class="va">treatment1id</span>,</span>
<span>    <span class="va">synergy_summaries</span><span class="op">$</span><span class="va">treatment2id</span>,</span>
<span>    sep<span class="op">=</span><span class="st">"_+_"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co">## Removing the two columns of treatment ids, and converting to a numerical matrix.</span></span>
<span><span class="va">synergy_summaries</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.matrix.html" class="external-link">data.matrix</a></span><span class="op">(</span><span class="va">synergy_summaries</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we can run our biomarker discovery. Note that we pass in the
treatment x sample matrix to the <code>sProfiles</code> argument, short
for sensitivity profiles. For the purposes of this demonstration, we
also subset the features we are computing over to 100 random genes,
simply to save time.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_exp_sigs</span> <span class="op">&lt;-</span> <span class="fu">drugSensitivitySig</span><span class="op">(</span><span class="va">NCI_ALMANAC_2017</span>,</span>
<span>  mDataType <span class="op">=</span> <span class="st">"rnaseq.comp"</span>,</span>
<span>  sProfiles <span class="op">=</span> <span class="va">synergy_summaries</span>,</span>
<span>  modeling.method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu">molecularProfiles</span><span class="op">(</span><span class="va">NCI_ALMANAC_2017</span>, <span class="st">"rnaseq.comp"</span><span class="op">)</span><span class="op">)</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  nthread <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing drug sensitivity signatures...</span></span></code></pre>
<p>The object returned by <code>drugSensitivitySig</code> is a 3D array,
with the first two dimensions of features X treatments, and the third
dimension returning computed statistics, the associated confidence
intervals, p values, and FDR correction.</p>
<p>One intersting (toy) example we find here is that expression of a
couple of genes are significantly correlated with antagonism (negative
Mean Bliss Score) for the combination of Vinblastine and Dasatinib.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">significant_vin_das</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  <span class="va">gene_exp_sigs</span><span class="op">[</span>, <span class="st">"Vinblastine_+_Dasatinib"</span>, <span class="op">]</span>,</span>
<span>  <span class="va">gene_exp_sigs</span><span class="op">[</span>, <span class="st">"Vinblastine_+_Dasatinib"</span>, <span class="st">"significant"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           estimate  n df significant       pvalue      lower      upper</span></span>
<span><span class="co">## NEUROD2 -0.5835669 42 30           1 0.0004550797 -0.7746336 -0.2948762</span></span>
<span><span class="co">## MOBP    -0.6246952 42 30           1 0.0001325058 -0.7992833 -0.3528649</span></span>
<span><span class="co">##                fdr</span></span>
<span><span class="co">## NEUROD2 0.02275399</span></span>
<span><span class="co">## MOBP    0.01325058</span></span></code></pre>
<p>To get a more global view of the results, we can also plot the
resulting object, which by default will display as a volcano plot. Note
below that the object returned is a normal ggplot2 plot, and therefore
we can add other annotations to it, such as the gene labels. Below, we
will add the labels for the two genes predicting antagonism between
Vinblastine and Dasatinib to the plot.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gene_exp_sigs</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">significant_vin_das</span><span class="op">[</span>, <span class="st">"estimate"</span><span class="op">]</span>,</span>
<span>                        y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">significant_vin_das</span><span class="op">[</span>, <span class="st">"pvalue"</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                        label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">significant_vin_das</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><img src="PGxWorkshop_files/figure-html/plotDrugSensSig-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="further-resources">Further Resources<a class="anchor" aria-label="anchor" href="#further-resources"></a>
</h2>
<p>We hope that this workshop leaves you feeling familiar with both the
data structures defined in <em><a href="https://bioconductor.org/packages/3.16/CoreGx" class="external-link">CoreGx</a></em> as
well as the analysis of drug combination experiments. Over time, our
group will be curating public drug combination studies and releasing
<code>PharmacoSet</code> objects through our <em><a href="https://orcestra.ca" class="external-link">orcestra.ca</a></em> platform, which will be
available for everyone to download and use in their own research. These
studies will use identifiers mapped to the same databases (Pubchem and
Cellosaurus) as already existing monotherapy PSets, allowing studies
combining monotherapy and combination response data.</p>
<p>Furthermore, on <em><a href="https://orcestra.ca" class="external-link">orcestra.ca</a></em> we are curating clinical
datasets combining genomic or transcriptomic profiling of cancer patient
samples with clinical drug response, mapped to the same identifiers and
processed as much as possible with the same pipelines. Much of the time,
clinical treatments are done in combination therapies, as standard of
care often involves more than 1 treatment. As this resource grows, it
will become possible to evaluate how well cell line combination
experiment results translate to clinical outcomes.</p>
<hr>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-holbeckNationalCancerInstitute2017" class="csl-entry">
Holbeck, Susan L., Richard Camalier, James A. Crowell, Jeevan Prasaad
Govindharajulu, Melinda Hollingshead, Lawrence W. Anderson, Eric Polley,
et al. 2017. <span>“The <span>National Cancer Institute ALMANAC</span>:
<span>A Comprehensive Screening Resource</span> for the
<span>Detection</span> of <span>Anticancer Drug Pairs</span> with
<span>Enhanced Therapeutic Activity</span>.”</span> <em>Cancer
Research</em> 77 (13): 3564–76. <a href="https://doi.org/10.1158/0008-5472.CAN-17-0489" class="external-link">https://doi.org/10.1158/0008-5472.CAN-17-0489</a>.
</div>
<div id="ref-yadavSearchingDrugSynergy2015" class="csl-entry">
Yadav, Bhagwan, Krister Wennerberg, Tero Aittokallio, and Jing Tang.
2015. <span>“Searching for <span>Drug Synergy</span> in <span>Complex
Dose</span> an <span>Interaction Potency Model</span>.”</span>
<em>Computational and Structural Biotechnology Journal</em> 13
(January): 504–13. <a href="https://doi.org/10.1016/j.csbj.2015.09.001" class="external-link">https://doi.org/10.1016/j.csbj.2015.09.001</a>.
</div>
</div>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<pre><code><span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ggplot2_3.3.6               data.table_1.14.2          </span></span>
<span><span class="co">##  [3] PharmacoGx_3.1.5            CoreGx_2.1.7               </span></span>
<span><span class="co">##  [5] SummarizedExperiment_1.27.1 Biobase_2.57.1             </span></span>
<span><span class="co">##  [7] GenomicRanges_1.49.0        GenomeInfoDb_1.33.3        </span></span>
<span><span class="co">##  [9] IRanges_2.31.0              S4Vectors_0.35.1           </span></span>
<span><span class="co">## [11] MatrixGenerics_1.9.1        matrixStats_0.62.0         </span></span>
<span><span class="co">## [13] BiocGenerics_0.43.1         pgxBioc2022Workshop_1.0.9  </span></span>
<span><span class="co">## [15] BiocStyle_2.25.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fgsea_1.23.0                colorspace_2.0-3           </span></span>
<span><span class="co">##   [3] ellipsis_0.3.2              rprojroot_2.0.3            </span></span>
<span><span class="co">##   [5] lsa_0.73.3                  XVector_0.37.0             </span></span>
<span><span class="co">##   [7] fs_1.5.2                    farver_2.1.1               </span></span>
<span><span class="co">##   [9] SnowballC_0.7.0             MultiAssayExperiment_1.23.4</span></span>
<span><span class="co">##  [11] DT_0.23                     fansi_1.0.3                </span></span>
<span><span class="co">##  [13] codetools_0.2-18            cachem_1.0.6               </span></span>
<span><span class="co">##  [15] knitr_1.39                  jsonlite_1.8.0             </span></span>
<span><span class="co">##  [17] magicaxis_2.2.14            cluster_2.1.3              </span></span>
<span><span class="co">##  [19] shinydashboard_0.7.2        shiny_1.7.2                </span></span>
<span><span class="co">##  [21] mapproj_1.2.8               BiocManager_1.30.18        </span></span>
<span><span class="co">##  [23] compiler_4.2.0              backports_1.4.1            </span></span>
<span><span class="co">##  [25] Matrix_1.4-1                fastmap_1.1.0              </span></span>
<span><span class="co">##  [27] bench_1.1.2                 limma_3.53.5               </span></span>
<span><span class="co">##  [29] cli_3.3.0                   later_1.3.0                </span></span>
<span><span class="co">##  [31] visNetwork_2.1.0            htmltools_0.5.3            </span></span>
<span><span class="co">##  [33] tools_4.2.0                 igraph_1.3.4               </span></span>
<span><span class="co">##  [35] gtable_0.3.0                glue_1.6.2                 </span></span>
<span><span class="co">##  [37] GenomeInfoDbData_1.2.8      reshape2_1.4.4             </span></span>
<span><span class="co">##  [39] RANN_2.6.1                  dplyr_1.0.9                </span></span>
<span><span class="co">##  [41] maps_3.4.0                  fastmatch_1.1-3            </span></span>
<span><span class="co">##  [43] Rcpp_1.0.9                  slam_0.1-50                </span></span>
<span><span class="co">##  [45] jquerylib_0.1.4             pkgdown_2.0.6              </span></span>
<span><span class="co">##  [47] vctrs_0.4.1                 BumpyMatrix_1.5.0          </span></span>
<span><span class="co">##  [49] xfun_0.31                   stringr_1.4.0              </span></span>
<span><span class="co">##  [51] mime_0.12                   lifecycle_1.0.1            </span></span>
<span><span class="co">##  [53] gtools_3.9.3                MASS_7.3-57                </span></span>
<span><span class="co">##  [55] zlibbioc_1.43.0             scales_1.2.0               </span></span>
<span><span class="co">##  [57] ragg_1.2.2                  promises_1.2.0.1           </span></span>
<span><span class="co">##  [59] relations_0.6-12            parallel_4.2.0             </span></span>
<span><span class="co">##  [61] RColorBrewer_1.1-3          sets_1.0-21                </span></span>
<span><span class="co">##  [63] yaml_2.3.5                  memoise_2.0.1              </span></span>
<span><span class="co">##  [65] gridExtra_2.3               downloader_0.4             </span></span>
<span><span class="co">##  [67] sass_0.4.2                  stringi_1.7.8              </span></span>
<span><span class="co">##  [69] highr_0.9                   NISTunits_1.0.1            </span></span>
<span><span class="co">##  [71] desc_1.4.1                  plotrix_3.8-2              </span></span>
<span><span class="co">##  [73] checkmate_2.1.0             caTools_1.18.2             </span></span>
<span><span class="co">##  [75] boot_1.3-28                 BiocParallel_1.31.10       </span></span>
<span><span class="co">##  [77] rlang_1.0.4                 pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [79] systemfonts_1.0.4           bitops_1.0-7               </span></span>
<span><span class="co">##  [81] pracma_2.3.8                evaluate_0.15              </span></span>
<span><span class="co">##  [83] lattice_0.20-45             purrr_0.3.4                </span></span>
<span><span class="co">##  [85] labeling_0.4.2              htmlwidgets_1.5.4          </span></span>
<span><span class="co">##  [87] tidyselect_1.1.2            plyr_1.8.7                 </span></span>
<span><span class="co">##  [89] magrittr_2.0.3              bookdown_0.27              </span></span>
<span><span class="co">##  [91] R6_2.5.1                    gplots_3.1.3               </span></span>
<span><span class="co">##  [93] generics_0.1.3              sm_2.2-5.7.1               </span></span>
<span><span class="co">##  [95] DelayedArray_0.23.0         withr_2.5.0                </span></span>
<span><span class="co">##  [97] pillar_1.8.0                RCurl_1.98-1.7             </span></span>
<span><span class="co">##  [99] tibble_3.1.8                crayon_1.5.1               </span></span>
<span><span class="co">## [101] KernSmooth_2.23-20          utf8_1.2.2                 </span></span>
<span><span class="co">## [103] rmarkdown_2.14              grid_4.2.0                 </span></span>
<span><span class="co">## [105] marray_1.75.0               piano_2.13.0               </span></span>
<span><span class="co">## [107] digest_0.6.29               xtable_1.8-4               </span></span>
<span><span class="co">## [109] httpuv_1.6.5                textshaping_0.3.6          </span></span>
<span><span class="co">## [111] celestial_1.4.6             coop_0.6-3                 </span></span>
<span><span class="co">## [113] munsell_0.5.0               bslib_0.4.0                </span></span>
<span><span class="co">## [115] shinyjs_2.1.0</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christopher Eeles, Petr Smirnov, Feifei Li, Benjamin Haibe-Kains.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
